---
title: "AZ-SciLife ExoDiab MS Proteomics analysis"
author: "Klev Diamanti, PhD"
date: "`r format(Sys.time(), '%A, %d %B %Y, %H:%M')`"
output:
  html_document:
    dev: 'svg'
---

```{r setup, echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
# set output directory
output.directory <- "../results/"
# set knitr options
knitr::opts_chunk$set(echo = TRUE, fig.path = output.directory)

# load libraries
suppressWarnings(suppressPackageStartupMessages(library(readxl)))
suppressWarnings(suppressPackageStartupMessages(library(knitr)))
suppressWarnings(suppressPackageStartupMessages(library(tidyverse)))
suppressWarnings(suppressPackageStartupMessages(library(tidyselect)))
suppressWarnings(suppressPackageStartupMessages(library(reshape2)))
suppressWarnings(suppressPackageStartupMessages(library(cowplot)))
suppressWarnings(suppressPackageStartupMessages(library(kableExtra)))
suppressWarnings(suppressPackageStartupMessages(library(gridExtra)))
suppressWarnings(suppressPackageStartupMessages(library(limma)))

## Other libraries that will be loaded but not attached
# data.table
# plyr
# dendextend
# magrittr
# qlcMatrix
# quantable
# ProteoMM
# DEqMS
# NormalyzerDE
# variancePartition
# pcaMethods
# corrplot
# BiocParallel
# ComplexHeatmap
# circlize
# Rtsne
# fgsea
# mgsub
# xlsx
# umap
# imputeLCMD
# impute
# ggdendro

# set java options to output files
options(java.parameters = "-Xmx8000m")
```

```{r setup.init, echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
## INPUT FILES
proteomics.file <- "../data/proteomics_random.xlsx"
covars.file <- "../covariates/covars_random.xlsx"
gene.id.file <- "../data/gsea_files/all.gene.ids.txt"

## TISSUES
tissues <- c("Fat", "Liver", "Muscle", "Pancreas", "Serum")
tissue.colors <- setNames(object = c('#814374ff', '#51a39dff', '#b7695cff', '#cdbb79ff', '#06425cff'), nm = tissues)

## MAIN PHENOTYPE
covars.pheno <- "pheno_3c"

## FILTERING QC
min.samples.threshold <- 0.8 # fraction of samples a protein was identified in
samples.to.remove.manually <- c("rand3408") # hardcoded removal of samples

## SIGNIFICANCE
# thresholds for pi-values
# thresh_pi <- 0.2572 # p = 0.2
# thresh_pi <- 0.4292 # p = 0.1
# thresh_pi <- 0.6274 # p = 0.05
# thresh_pi <- 1.1733 # p = 0.01
# thresh_pi <- 2.1281 # p = 0.001
significance.signs <- data.frame(sign = c("***", "**", "*", "#", "##", "NS"),
                                 p.q = c(0.001, 0.01, 0.05, 0.1, 0.2, 1),
                                 pi = c(2.1281, 1.1733, 0.6274, 0.4292, 0.2572, 0), stringsAsFactors = F)
significance.threshold.p.q.relaxed <- 0.1
significance.threshold.p.q.strict <- 0.05
significance.threshold.p.q.strictest <- 0.001
significance.threshold.pi <- significance.signs %>%
  filter(p.q <= significance.threshold.p.q.relaxed) %>%
  slice(n()) %>%
  pull(pi)

## CONTRASTS TO CONSIDER IN THE DIFFERENTIAL ANALYSIS (DE)
within.tissues.contrasts <- setNames(object = c("t2dVSctrl", "t2dVSpd", "pdVSctrl", "t2dVSctrl_pd", "pd_t2dVSctrl"),
                                     nm = c("T2D v CTRL", "T2D v PD", "PD v CTRL", "T2D v CTRL+PD", "PD+T2D v CTRL"))

## FILES WITH GO TERMS AND PATHWAYS
biological.terms.files <- data.frame(type = c("GO", "GO", "GO", "Pathway"),
                                     subtype = c("bp", "cc", "mf", ""),
                                     file = c("../data/gsea_files/gmt_files/c5.go.bp.v7.2.symbols.kd.gmt", "../data/gsea_files/gmt_files/c5.go.cc.v7.2.symbols.kd.gmt", "../data/gsea_files/gmt_files/c5.go.mf.v7.2.symbols.kd.gmt", "../data/gsea_files/gmt_files/c2.cp.v7.2.symbols.kd.gmt"),
                                     stringsAsFactors = F) %>%
  mutate(combo = paste(type, subtype, sep = ""))
GOslim.file <- "../data/gsea_files/obo_files/goslim_analysis-ready.tsv"
GOall.file <- "../data/gsea_files/gmt_files/c2cpANDc5go.v7.2.symbols.kd.gmt"

## PARAMETERS FOR GO AND PATHWAY ENRICHMENT ANALYSIS
min.num.of.proteins.in.term <- 5
max.num.of.proteins.term <- 300
num.of.permutations <- 1e+06

## SEED
seed <- 1
set.seed(seed = seed)

# OS-SPECIFIC ARGUMENTS AND THREADS
switch(Sys.info()[['sysname']],
       Windows = {
         num.of.cores.total <- parallel::detectCores() - 1
         bash.cmd.prefix <- "wsl " },
       Linux   = {
         num.of.cores.total <- parallel::detectCores() - 1
         bash.cmd.prefix <- "" },
       Darwin  = {
         num.of.cores.total <- parallel::detectCores() - 1
         bash.cmd.prefix <- "" })
if (num.of.cores.total <= length(tissues)) {
  num.of.cores.tissues <- num.of.cores.total
} else {
  num.of.cores.tissues <- 5
}

## WRITE TO EXCEL FUNCTION
write.to.excel <- function(df_, file_, sheet_, append_) {
  df_ <- as.data.frame(df_, check.names = F, fix.empty.names = F)
  rownames(df_) <- NULL
  xlsx::write.xlsx2(x = df_, file = file_, sheetName = sheet_, col.names = TRUE, row.names = FALSE, append = append_)
}

# HELPER FUNCTION TO COMBINE ALL DATASETS FOR ALL TISSUES IN ONE MATRIX
merge.full.joins <- function(list.of.df_, by_) {
  merge.tmp <- list.of.df_[[1]]
  for (i in 2:length(list.of.df_)) {
    merge.tmp <- full_join(merge.tmp, list.of.df_[[i]], by = by_)
  }
  return(merge.tmp)
}

merge.unique.proteins.dia <- function(dia.all_) {
  tmp.names <- names(dia.all_)
  dia.all_ <- lapply(seq_along(dia.all_), function(i) {
    tmp.dia.all_ <- dia.all_[[i]]
    colnames(tmp.dia.all_) <- c("AZpID", paste(names(dia.all_)[[i]], tail(colnames(tmp.dia.all_), -1), sep = "."))
    return(tmp.dia.all_)
  })
  names(dia.all_) <- tmp.names
  
  return(merge.full.joins(list.of.df_ = dia.all_, by_ = 'AZpID'))
}

merge.unique.samples.dia <- function(dia.all_) {
  merge.unique.samples.dia.help <- function(dia_, tissue_) {
    dia.t_ <- data.table::transpose(l = dia_ %>% column_to_rownames("AZpID")) %>%
      as.data.frame()
    colnames(dia.t_) <- dia_$AZpID
    dia.t_ <- dia.t_ %>%
      mutate(id = tail(colnames(dia_), -1)) %>%
      select(id, everything())
    colnames(dia.t_) <- c(head(colnames(dia.t_), 1), paste(tail(colnames(dia.t_), -1), ".", tissue_, sep = ""))
    return(dia.t_)
  }
  dia.t.all_ <- lapply(seq_along(dia.all_), function(i) merge.unique.samples.dia.help(dia_ = dia.all_[[i]], tissue_ = names(dia.all_)[i]))
  return(merge.full.joins(list.of.df_ = dia.t.all_, by_ = 'id'))
}
```

# Analyzing mass spectrometry (MS) proteomics for 5 tissues
- Visceral Adipose Tissue (VAT) - herein Fat
- Liver
- Skeletal Muscle - herein Muscle
- Pancreatic Islets - herein Pancreas
- Blood Serum - herein Serum

# Reading the data

## Reading clinical variables

Reading the random clinical variables. Available anthropometric data include _age_, _sex_, _body mass index (bmi)_, _blood group_, _% of purity of islets in sample_ and _cold ischemic time (cit)_. Additional clinical variables related to type 2 diabetes (T2D) include _hemoglobin A1c (hba1c)_, _glucose-stimulated insulin secretion (GSIS)_ in pancreatic islets, and the discrete T2D phenotype _pheno_3c_ of _control (CTRL)_, _pre-diabetes (PD)_ and _T2D_. We introduce 2 additional columns _pehno_ctrl_pd_ and _pheno_pd_t2d_ that merge the group of PD with CTRL and T2D, respectively.

The table below shows distribution of samples across discrete phenotypes in columns _CTRL_, _PD_, _T2D_ and _NA_, where _NA_ shows samples with non available phenotype information. Columns starting with _mean Â± sd_ contain the average and the standard deviation for the corresponding row variable in each phenotypic group. The columns _P (CTRL v PD)_, _P (CTRL v T2D)_, _P (PD v T2D)_, _P (CTRL v PD+T2D)_ and _P (PD+T2D v CTRL)_ show the statistical significance measured by a Mann-Whitney U test for the groups in parentheses, while the column _P (CTRL v PD v T2D)_ shows statistical significance across all 3 phenotypic groups in _pheno_3c_ measured by a Kruskal-Wallis test. The columns _P (bmi)_, _P (hba1c)_ and _P (gsis)_ represent associations measured by a linear regression. The row for _sex_ shows the distribution of sexes across phenotypic groups.

```{r read.covars, echo=FALSE, warning=FALSE, message=FALSE, results='asis'}
# read covars file
covars <- read_excel(path = covars.file, sheet = 1) %>%
  mutate_all(as.character) %>%
  map_df(~parse_guess(.))

# assign predefined colors to phenotypes
phenotypes.colors <- data.frame(names = c(sort(unique(c(covars$pheno_3c, covars$pheno_ctrl_pd, covars$pheno_pd_t2d))), "CTRL_PD__T2D", "PD_T2D__CTRL"),
                                colors = c('#dcd0c0ff', '#bbbac6ff', '#e2e2e2ff', '#90afc5ff', '#b7ceceff', '#ffffffff', '#ffffffff'), stringsAsFactors = F) %>%
  mutate(class = c("pheno_3c", "pheno_ctrl_pd", "pheno_3c", "pheno_pd_t2d", "pheno_3c", "pheno_ctrl_pd", "pheno_pd_t2d")) %>%
  bind_rows(data.frame(names = c("CTRL", "T2D"), colors = c('#dcd0c0ff', '#b7ceceff'), class = c("pheno_pd_t2d", "pheno_ctrl_pd"), stringsAsFactors = F))

# mark variables to use as response for various analyses
numeric.response <- c("bmi", "hba1c", "gsis")
discrete.response <- c("pheno_3c", "pheno_ctrl_pd", "pheno_pd_t2d")
exclude.cols <- c("id", "blood.group")

# mark variables account for in the various analyses
numeric.covars <- sapply(covars, class) %>%
  as.data.frame() %>%
  rename("class" = ".") %>%
  rownames_to_column("column") %>%
  filter(!(column %in% exclude.cols) & class == "numeric") %>%
  pull(column)

discrete.covars <- sapply(covars, class) %>%
  as.data.frame() %>%
  rename("class" = ".") %>%
  rownames_to_column("column") %>%
  filter(!(column %in% c(exclude.cols, discrete.response)) & class == "character") %>%
  pull(column)

# linear regressions using the numeric responses and all other covars as predictors
numeric.response.p <- bind_rows(lapply(numeric.response, function(resp.var){
  tmp.resp <- pull(covars, resp.var)
  select(covars, all_of(c(numeric.covars, discrete.covars, discrete.response))) %>%
    map(function(tmp.pred) {
      if (identical(tmp.pred, tmp.resp)) {
        NA
      } else {
        modelobject <- lm(tmp.resp ~ tmp.pred)
        f <- summary(modelobject)$fstatistic
        p <- pf(f[1],f[2],f[3],lower.tail=F)
        attributes(p) <- NULL
        formatC(p, format = "e", digits = 1)
      }})})) %>%
  mutate(Numeric.Response = numeric.response) %>%
  column_to_rownames("Numeric.Response") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Variable")
colnames(numeric.response.p) <- c(colnames(numeric.response.p)[1], paste("P (", tail(colnames(numeric.response.p), -1), ")", sep = ""))

# Mann_whitney U test for pheno_3c discrete responses
discrete.response.qualitative.3c.num <- select(covars, pheno_3c, all_of(numeric.covars)) %>%
  gather(key = Variable, value = value, -pheno_3c) %>% 
  group_by(pheno_3c, Variable) %>% 
  summarise(value = list(value)) %>% 
  spread(pheno_3c, value) %>% 
  rename("mNA" = "<NA>") %>% 
  group_by(Variable) %>%
  mutate(num_CTRL = length(unlist(CTRL)),
         num_PD = length(unlist(PD)),
         num_T2D = length(unlist(T2D)),
         num_mNA = length(unlist(mNA)),
         mean_CTRL = mean(x = unlist(CTRL), na.rm = T),
         sd_CTRL = sd(x = unlist(CTRL), na.rm = T),
         mean_PD = mean(x = unlist(PD), na.rm = T),
         sd_PD = sd(x = unlist(PD), na.rm = T),
         mean_T2D = mean(x = unlist(T2D), na.rm = T),
         sd_T2D = sd(x = unlist(T2D), na.rm = T),
         mean_mNA = mean(x = unlist(mNA), na.rm = T),
         sd_mNA = sd(x = unlist(mNA), na.rm = T)) %>%
  mutate_at(.vars = vars(num_CTRL:sd_mNA), .funs = list(~(ifelse(is.infinite(.), NA, ifelse(is.nan(.), NA, .))))) %>%
  mutate(mean_sd_CTRL = paste(round(x = mean_CTRL, digits = 1), "?", round(x = sd_CTRL, digits = 1)),
         mean_sd_PD = paste(round(x = mean_PD, digits = 1), "?", round(x = sd_PD, digits = 1)),
         mean_sd_T2D = paste(round(x = mean_T2D, digits = 1), "?", round(x = sd_T2D, digits = 1)),
         mean_sd_mNA = ifelse(is.na(mean_mNA) & is.na(sd_mNA), NA, paste(round(x = mean_mNA, digits = 1), "?", round(x = sd_mNA, digits = 1)))) %>%
  mutate(p_CTRL_PD_T2D = formatC(x = kruskal.test(x = c(unlist(CTRL), unlist(PD), unlist(T2D)),
                                                  g = as.factor(c(rep("CTRL", length(unlist(CTRL))), rep("PD", length(unlist(PD))), rep("T2D", length(unlist(T2D))))))$p.value,
                                 format = "e", digits = 1),
         p_CTRL_PD = formatC(x = pairwise.wilcox.test(x = c(unlist(CTRL), unlist(PD), unlist(T2D)),
                                                      g = c(rep("CTRL", length(unlist(CTRL))), rep("PD", length(unlist(PD))), rep("T2D", length(unlist(T2D)))),
                                                      p.adjust.method = "none", paired = FALSE, na.action = na.omit)$p.value[1,1],
                             format = "e", digits = 1),
         p_CTRL_T2D = formatC(x = pairwise.wilcox.test(x = c(unlist(CTRL), unlist(PD), unlist(T2D)),
                                                       g = c(rep("CTRL", length(unlist(CTRL))), rep("PD", length(unlist(PD))), rep("T2D", length(unlist(T2D)))),
                                                       p.adjust.method = "none", paired = FALSE, na.action = na.omit)$p.value[2,1],
                              format = "e", digits = 1),
         p_PD_T2D = formatC(x = pairwise.wilcox.test(x = c(unlist(CTRL), unlist(PD), unlist(T2D)),
                                                     g = c(rep("CTRL", length(unlist(CTRL))), rep("PD", length(unlist(PD))), rep("T2D", length(unlist(T2D)))),
                                                     p.adjust.method = "none", paired = FALSE, na.action = na.omit)$p.value[2,2],
                            format = "e", digits = 1)) %>%
  select(Variable, starts_with("num_"), starts_with("mean_sd"), starts_with("p_"))

# Fraction of males and females
discrete.response.qualitative.3c.disc <- select(covars, pheno_3c, all_of(discrete.covars)) %>%
  gather(key = Variable, value = value, -pheno_3c) %>% 
  group_by(pheno_3c, Variable) %>% 
  summarise(value = list(value)) %>% 
  spread(pheno_3c, value) %>% 
  rename("mNA" = "<NA>") %>% 
  group_by(Variable) %>%
  mutate(across(.fns = list(num = ~length(unlist(.)),
                            mean_sd = ~paste(as.data.frame(table(factor(x = unlist(.)), useNA = "ifany")) %>% arrange(Var1) %>% mutate(Str = paste(Freq, Var1)) %>% pull(Str), collapse = " / ")),
                .names = "{fn}_{col}"),
         p_CTRL_PD_T2D = NA,
         p_CTRL_PD = NA,
         p_CTRL_T2D = NA,
         p_PD_T2D = NA) %>%
  select(Variable, starts_with("num_"), starts_with("mean_sd"), starts_with("p_"))

# combine results from above
discrete.response.qualitative.3c <- bind_rows(discrete.response.qualitative.3c.num, discrete.response.qualitative.3c.disc)
rm(discrete.response.qualitative.3c.num, discrete.response.qualitative.3c.disc)

discrete.response.qualitative.ctrl_pd.num <- select(covars, pheno_ctrl_pd, all_of(numeric.covars)) %>%
  gather(key = Variable, value = value, -pheno_ctrl_pd) %>% 
  group_by(pheno_ctrl_pd, Variable) %>% 
  summarise(value = list(value)) %>% 
  spread(pheno_ctrl_pd, value) %>%
  group_by(Variable) %>%
  mutate(num_CTRL_PD = length(unlist(CTRL_PD)),
         mean_CTRL_PD = mean(x = unlist(CTRL_PD), na.rm = T),
         sd_CTRL_PD = sd(x = unlist(CTRL_PD), na.rm = T),
         mean_sd_CTRL_PD = paste(round(x = mean_CTRL_PD, digits = 1), "?", round(x = sd_CTRL_PD, digits = 1))) %>%
  mutate(p_CTRL_PD__T2D = formatC(x = pairwise.wilcox.test(x = c(unlist(CTRL_PD), unlist(T2D)),
                                                           g = c(rep("CTRL_PD", length(unlist(CTRL_PD))), rep("T2D", length(unlist(T2D)))),
                                                           p.adjust.method = "none", paired = FALSE, na.action = na.omit)$p.value[1,1],
                                  format = "e", digits = 1)) %>%
  select(Variable, starts_with("num_"), starts_with("mean_sd"), starts_with("p_"))

discrete.response.qualitative.ctrl_pd.disc <- select(covars, pheno_ctrl_pd, all_of(discrete.covars)) %>%
  gather(key = Variable, value = value, -pheno_ctrl_pd) %>% 
  group_by(pheno_ctrl_pd, Variable) %>% 
  summarise(value = list(value)) %>% 
  spread(pheno_ctrl_pd, value) %>%
  group_by(Variable) %>%
  mutate(across(.fns = list(num = ~length(unlist(.)),
                            mean_sd = ~paste(as.data.frame(table(factor(x = unlist(.)), useNA = "ifany")) %>% arrange(Var1) %>% mutate(Str = paste(Freq, Var1)) %>% pull(Str), collapse = " / ")),
                .names = "{fn}_{col}"),
         p_CTRL_PD__T2D = NA) %>%
  select(colnames(discrete.response.qualitative.ctrl_pd.num))

discrete.response.qualitative.ctrl_pd <- bind_rows(discrete.response.qualitative.ctrl_pd.num, discrete.response.qualitative.ctrl_pd.disc)
rm(discrete.response.qualitative.ctrl_pd.num, discrete.response.qualitative.ctrl_pd.disc)

discrete.response.qualitative.pd_t2d.num <- select(covars, pheno_pd_t2d, all_of(numeric.covars)) %>%
  gather(key = Variable, value = value, -pheno_pd_t2d) %>% 
  group_by(pheno_pd_t2d, Variable) %>% 
  summarise(value = list(value)) %>% 
  spread(pheno_pd_t2d, value) %>%
  group_by(Variable) %>%
  mutate(num_PD_T2D = length(unlist(PD_T2D)),
         mean_PD_T2D = mean(x = unlist(PD_T2D), na.rm = T),
         sd_PD_T2D = sd(x = unlist(PD_T2D), na.rm = T),
         mean_sd_PD_T2D = paste(round(x = mean_PD_T2D, digits = 1), "?", round(x = sd_PD_T2D, digits = 1))) %>%
  mutate(p_PD_T2D__CTRL = formatC(x = pairwise.wilcox.test(x = c(unlist(PD_T2D), unlist(CTRL)),
                                                           g = c(rep("PD_T2D", length(unlist(PD_T2D))), rep("CTRL", length(unlist(CTRL)))),
                                                           p.adjust.method = "none", paired = FALSE, na.action = na.omit)$p.value[1,1],
                                  format = "e", digits = 1)) %>%
  select(Variable, starts_with("num_"), starts_with("mean_sd"), starts_with("p_"))

discrete.response.qualitative.pd_t2d.disc <- select(covars, pheno_pd_t2d, all_of(discrete.covars)) %>%
  gather(key = Variable, value = value, -pheno_pd_t2d) %>% 
  group_by(pheno_pd_t2d, Variable) %>% 
  summarise(value = list(value)) %>% 
  spread(pheno_pd_t2d, value) %>%
  group_by(Variable) %>%
  mutate(across(.fns = list(num = ~length(unlist(.)),
                            mean_sd = ~paste(as.data.frame(table(factor(x = unlist(.)), useNA = "ifany")) %>% arrange(Var1) %>% mutate(Str = paste(Freq, Var1)) %>% pull(Str), collapse = " / ")),
                .names = "{fn}_{col}"),
         p_PD_T2D__CTRL = NA) %>%
  select(colnames(discrete.response.qualitative.pd_t2d.num))

discrete.response.qualitative.pd_t2d <- bind_rows(discrete.response.qualitative.pd_t2d.num, discrete.response.qualitative.pd_t2d.disc)
rm(discrete.response.qualitative.pd_t2d.num, discrete.response.qualitative.pd_t2d.disc)

discrete.response.qualitative <- left_join(discrete.response.qualitative.3c, discrete.response.qualitative.ctrl_pd, by = 'Variable') %>%
  left_join(discrete.response.qualitative.pd_t2d, by = 'Variable') %>%
  select(Variable, num_CTRL, num_PD, num_T2D, num_mNA, mean_sd_CTRL, mean_sd_PD, mean_sd_CTRL_PD, mean_sd_T2D, mean_sd_PD_T2D, mean_sd_mNA, p_CTRL_PD, p_PD_T2D, p_CTRL_T2D, p_CTRL_PD_T2D, p_CTRL_PD__T2D, p_PD_T2D__CTRL)
colnames(discrete.response.qualitative) <- c(head(colnames(discrete.response.qualitative), 1), unlist(lapply(strsplit(x = tail(colnames(discrete.response.qualitative), -1), split = "_", fixed = T), function(i){
  if (i[1] == "num") {
    paste(tail(i, -1), collapse = "-")
  } else if (i[1] == "mean") {
    paste(i[1], " ? ", i[2], " (", paste(tail(i, -2), collapse = "-"), ")", sep = "")
  } else if (i[1] == "p") {
    paste("P (", paste(tail(i, -1), collapse = "-"), ")", sep = "")
  }
})))
rm(discrete.response.qualitative.3c, discrete.response.qualitative.ctrl_pd, discrete.response.qualitative.pd_t2d)

covars.stats <- full_join(discrete.response.qualitative, numeric.response.p, by = 'Variable') %>%
  ungroup() %>%
  mutate(Variable = case_when(
    Variable == "age" ~ "Age",
    Variable == "bmi" ~ "BMI",
    Variable == "cit.min" ~ "CIT (min)",
    Variable == "days.in.icu" ~ "Days in ICU",
    Variable == "gsis" ~ "GSIS",
    Variable == "hba1c" ~ "HbA1c",
    Variable == "islets.purity" ~ "Islets purity",
    Variable == "antidiuretics" ~ "Antidiuretics",
    Variable == "antiepileptic" ~ "Antiepileptics",
    Variable == "antiinfectives" ~ "Antiinfectives",
    Variable == "bba" ~ "Beta blocking agents",
    Variable == "bsps" ~ "Blood substitutes and perfusion solutions",
    Variable == "cardiac.therapy" ~ "Cardiac therapy",
    Variable == "diuretics" ~ "Diuretics",
    Variable == "glucocorticoids" ~ "Glucocorticoids",
    Variable == "insulin" ~ "Insulin",
    Variable == "sex" ~ "Sex",
    Variable == "pheno_3c" ~ "CTRL v PD v T2D",
    Variable == "pheno_ctrl_pd" ~ "CTRL+PD v T2D",
    Variable == "pheno_pd_t2d" ~ "CTRL v PD+T2D",
    TRUE ~ "TMP")) %>%
  mutate_all(as.character) %>%
  mutate_all(.funs = list(~replace_na(., "-"))) %>%
  mutate_all(.funs = list(~gsub("?", "Â±", ., fixed = T)))
covars.stats <- magrittr::set_colnames(covars.stats, gsub("mNA", "NA", x = gsub("?", "Â±", x = gsub("-", " v ", x = gsub("--", "+", x = colnames(covars.stats), fixed = T), fixed = T), fixed = T), fixed = T))

if (file.exists(paste(output.directory, "covars.xlsx", sep = ""))) {
  file.remove(paste(output.directory, "covars.xlsx", sep = ""))
}
write.to.excel(df_ = covars, file_ = paste(output.directory, "covars.xlsx", sep = ""), sheet_ = "covars", append_ = FALSE)
write.to.excel(df_ = covars.stats, file_ = paste(output.directory, "covars.xlsx", sep = ""), sheet_ = "stats", append_ = TRUE)

rm(numeric.response, discrete.response, exclude.cols, numeric.covars, discrete.covars, discrete.response.qualitative, numeric.response.p)

kable(covars.stats) %>% 
  kable_styling(full_width = T, bootstrap_options = "striped", font_size = 8.5) %>%
  column_spec(1, bold = T, border_right = T)

rm(covars.stats)
```

_NB:_ there are a few not available (NA) values for some clinical variables.

## Reading, pre-processing and shaping the MS proteomics data

Read raw MS proteomics data. Create an index that contains the collection of unique custom identifiers for each protein, the UniProtKB identifier, the gene name and a brief description of it.

_NB:_ UniProtKB identifiers are resolved so that unique ones are maintained for downstream analysis. This is based on the order of the identifiers provided in the input file and whether they exist in the universe of identifiers from the GO terms and the pathways. For detailed description see the manuscript and explore the code

Pre-defined samples are filtered out, while proteins are filtered if <=`r paste(min.samples.threshold * 100, "%", sep = "")` of them is missing from samples.

The final object is a large list that contains all the data that allow the downstream analysis.

```{r reading.the.data.funs, echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
# read input files
read.dia <- function(tissue_, file_) {
  dia_ <- read_excel(path = file_, sheet = tissue_)
  
  unique.tissue <- unique(unlist(lapply(strsplit(x = tail(colnames(dia_), -3), split = "_", fixed = T), function(x) tail(x = head(x = x, n = 2), n = 1))))
  if (length(unique.tissue) > 1 || unique.tissue != tissue_) {
    stop("Error: No unique tissue for colnames or tissue from colnames does not match given tissue name!")
  }
  
  colnames(dia_) <- c("Protein_Accession", "Gene", "Protein_Description", unlist(lapply(strsplit(x = tail(colnames(dia_), -3), split = "_", fixed = T), head, 1)))
  
  return(mutate(dia_, AZpID = ifelse(row_number() < 10, paste(tissue_, ".AZp000", row_number(), sep = ""),
                                    ifelse(row_number() < 100, paste(tissue_, ".AZp00", row_number(), sep = ""),
                                           ifelse(row_number() < 1000, paste(tissue_, ".AZp0", row_number(), sep = ""), paste(tissue_, ".AZp", row_number(), sep = ""))))) %>%
           select(AZpID, everything()))
}

# create index for all proteins
get.dia.index <- function(dia_, tissue_) {
  return(dia_ %>%
           select(AZpID, Gene, Protein_Accession, Protein_Description) %>%
           rename(!!paste("AZpID", tissue_, sep = ".") := "AZpID",
                  !!paste("Gene", tissue_, sep = ".") := "Gene",
                  !!paste("Protein_Description", tissue_, sep = ".") := "Protein_Description"))
}

# shape proteomics measurements for downstream analysis
prepare.dia <- function(dia_, tissue_, dia.index_) {
  azpid.name <- paste("AZpID", tissue_, sep = ".")
  return(dia_ %>%
           rename(!!azpid.name := "AZpID") %>%
           left_join(select(dia.index_, AZpID, all_of(azpid.name)), by = azpid.name) %>%
           select(AZpID, everything()) %>%
           select(-all_of(azpid.name), -Gene, -Protein_Accession, -Protein_Description) %>%
           mutate_all(as.character) %>%
           map_df(~parse_guess(.)))
}

# resolve multiplicity of Gene names
unique.gene.id <- function(azpid.index_) {
  list.of.geneid <- read.delim(file = gene.id.file, header = F, sep = "\t", stringsAsFactors = F) %>%
    pull(V1)
  
  azpid.to.remove <- azpid.index_ %>%
    filter(is.na(Gene) | nchar(Gene) == 0) %>%
    pull(AZpID)
  
  spread.azpid.index <- azpid.index_ %>%
    filter(nchar(Gene) > 0) %>%
    filter(!is.na(Gene)) %>%
    separate_rows(Gene, sep = ";") %>%
    mutate(in.list = ifelse(Gene %in% list.of.geneid, "YES", "NO")) %>%
    group_by(AZpID) %>%
    filter(nchar(Gene) > 0) %>%
    filter(!is.na(Gene)) %>%
    distinct(Gene, .keep_all = T) %>%
    mutate(ncol = n(),
           indx = row_number())
  
  id.in.list <- spread.azpid.index %>%
    filter(in.list == "YES") %>%
    slice(1) %>%
    ungroup()
  
  id.in.list.add <- spread.azpid.index %>%
    filter_at(.vars = vars(in.list), all_vars(. == "NO")) %>%
    ungroup() %>%
    filter(!(AZpID %in% id.in.list$AZpID)) %>%
    separate_rows(Gene, sep = "-") %>%
    mutate(in.list = ifelse(Gene %in% list.of.geneid, "YES", "NO")) %>%
    group_by(AZpID) %>%
    filter(nchar(Gene) > 0) %>%
    filter(!is.na(Gene)) %>%
    distinct(Gene, .keep_all = T) %>%
    mutate(ncol = n(),
           indx = row_number()) %>%
    filter(in.list == "YES") %>%
    slice(1) %>%
    ungroup()
  
  id.not.in.list <- spread.azpid.index %>%
    filter_at(.vars = vars(in.list), all_vars(. == "NO")) %>%
    slice(1) %>%
    ungroup() %>%
    filter(!(AZpID %in% c(id.in.list$AZpID, id.in.list.add$AZpID)))
  
  unique.azpid <- bind_rows(id.in.list, id.in.list.add, id.not.in.list) %>%
    arrange(AZpID) %>%
    select(-in.list, -ncol, -indx)
  
  duplicated.gene.id <- unique.azpid %>%
    arrange(Gene, AZpID) %>%
    group_by(Gene) %>%
    mutate(ncol = n()) %>%
    filter(ncol > 1) %>%
    mutate(tmp.AZpID = AZpID[1]) %>%
    ungroup()
  
  azpid.index <- bind_rows(unique.azpid %>% anti_join(select(duplicated.gene.id, AZpID), by = 'AZpID'),
                           unique.azpid %>% inner_join(select(duplicated.gene.id, tmp.AZpID) %>% distinct(tmp.AZpID), by = c("AZpID" = "tmp.AZpID")))
  
  return(list(azpindex = azpid.index, duplicated = duplicated.gene.id, toremove = azpid.to.remove))
}

# filter proteins based on the % of samples the protein was identified on
filter.proteins <- function(dia_, min.samples.thresh_) {
  keep.overall <- dia_ %>%
             column_to_rownames("AZpID") %>%
             transpose() %>%
             lapply(., function(i) 1 - (sum(is.na(i)) / length(i))) %>%
             unlist() %>%
             enframe(., value = "V1") %>%
             mutate(AZpID = dia_$AZpID) %>%
             filter(V1 >= min.samples.thresh_)
  return(filter(dia_, AZpID %in% keep.overall$AZpID))
}

# filter pre-defined samples
filter.samples <- function(dia_, samples.to.remove.manually_) {
  keep.me <- tail(colnames(dia_), -1)
  keep.me <- keep.me[!(keep.me %in% samples.to.remove.manually_)]
  return(dia_ %>% select(AZpID, all_of(keep.me)))
}

# update index so that the custom identifiers refer to the same protein across tissues
update.index <- function(dia.index.all_, dia.filtered_, tissues_) {
  s <- lapply(tissues_, function(i) dia.index.all_ %>%
                select(AZpID, all_of(paste("AZpID", i, sep = "."))) %>%
                magrittr::set_names(c("AZpID", "tissue")) %>%
                mutate(!!i := ifelse(AZpID %in% dia.filtered_[[i]]$AZpID, tissue, NA)) %>%
                select(-tissue))
  
  si <- plyr::join_all(dfs = s, by = 'AZpID', type = 'left') %>%
    as_tibble()
  colnames(si) <- c(head(colnames(si), 1), paste("AZpID", tail(colnames(si), -1), sep = "."))
  
  return(left_join(select(dia.index.all_, AZpID, Gene, Protein_Accession, Protein_Description), si, by = 'AZpID'))
}

# average proteins that have identified twice within the same tissue
average.duplicate.protein.accessions <- function(dia_, dpa_) {
  average.of.duplicated <- dia_ %>%
    inner_join(select(dpa_, AZpID, tmp.AZpID), by = 'AZpID') %>%
    arrange(tmp.AZpID) %>%
    group_by(tmp.AZpID) %>%
    summarise_at(.vars = vars(starts_with("p")), .funs = list(~mean(., na.rm = T))) %>%
    mutate_all(.funs = list(~ifelse(is.nan(.), NA, .))) %>%
    rename("AZpID" = "tmp.AZpID")
  
  return(dia_ %>%
           anti_join(select(dpa_, AZpID), by = 'AZpID') %>%
           bind_rows(average.of.duplicated))
}
```

```{r reading.the.data, echo=FALSE, warning=FALSE, message=FALSE, results='hide'}
dia <- list() # dia is the list that will keep all the data
dia$raw <- lapply(tissues, read.dia, proteomics.file) # read the raw data for all the tissues from the input excel file
names(dia$raw) <- tissues # name the list of raw data according to the tissues

# make the index for Gene names, protein accessions etc.
dia$index <- list() # create an index list that will keep matches of IDs to protein accessions and gene names
dia$index$all <- lapply(seq_along(dia$raw), function(i) get.dia.index(dia_ = dia$raw[[i]], tissue_ = names(dia$raw)[i])) # retrieve all protein IDs
names(dia$index$all) <- tissues #name the list
dia$index$all <- plyr::join_all(dfs = dia$index$all, by = 'Protein_Accession', type = 'full') %>%
  unite(Gene, starts_with("Gene"), sep = '|', na.rm = T, remove = F) %>%
  unite(Protein_Description, starts_with("Protein_Description"), sep = '|', na.rm = T, remove = F) %>%
  mutate(Gene = unlist(lapply(lapply(strsplit(x = Gene, split = "|", fixed = T), unique), paste, collapse = "|")),
         Protein_Description = unlist(lapply(lapply(strsplit(x = Protein_Description, split = "|", fixed = T), unique), paste, collapse = "|")),
         AZpID = case_when(
           row_number() < 10 ~ paste("AZp000", row_number(), sep = ""),
           row_number() < 100 ~ paste("AZp00", row_number(), sep = ""),
           row_number() < 1000 ~ paste("AZp0", row_number(), sep = ""),
           TRUE ~ paste("AZp", row_number(), sep = ""))) %>%
  select(AZpID, Gene, Protein_Accession, Protein_Description, starts_with("AZpID.")) %>%
  mutate_at(.vars = vars(starts_with("AZpID.")), .funs = list(Present = ~ifelse(!is.na(.), "X", NA))) %>%
  rename_at(.vars = vars(ends_with("Present")), .funs = list(~paste("Present", unlist(lapply(strsplit(x = unlist(lapply(strsplit(x = ., split = "_", fixed = T), head, 1)), split = ".", fixed = T), function(x) paste(tail(x, 1), collapse = "."))), sep = "."))) %>%
  as_tibble()
dia$index$AZpID <- dia$index$all %>% select(AZpID, Gene, Protein_Accession, Protein_Description, starts_with("AZpID."))

# prepare the final set of raw data in a format proteins x samples
dia$raw <- lapply(seq_along(dia$raw), function(i) prepare.dia(dia_ = dia$raw[[i]], tissue_ = names(dia$raw)[i], dia.index_ = dia$index$AZpID))
names(dia$raw) <- tissues

# fix unique protein accessions
tmp.azpid <- unique.gene.id(azpid.index_ = dia$index$AZpID)
dia$index$AZpID <- tmp.azpid$azpindex
dia$index$all <- dia$index$all %>%
  select(-Protein_Accession) %>%
  inner_join(select(dia$index$AZpID, AZpID, Protein_Accession), by = 'AZpID') %>%
  select(AZpID:Gene, Protein_Accession, everything())
dia$index$index <- dia$index$all %>% select(AZpID, Gene, Protein_Accession, Protein_Description)
dia$index$Present <- dia$index$all %>% select(AZpID, Gene, Protein_Accession, Protein_Description, starts_with("Present"))

# filter the data set for proteins in min.samples.threshold
dia$raw.filtered <- lapply(dia$raw, filter.proteins, min.samples.thresh_ = min.samples.threshold)

# filter samples with large deviations from the average median and samples that we know are outliers
dia$raw.filtered <- lapply(dia$raw.filtered, filter.samples, samples.to.remove.manually_ = samples.to.remove.manually)

# remove the temporary AZpID per tissue we set when we first read the data into the unique global IDs
dia$index$all <- dia$index$all %>%
  mutate_at(.vars = vars(starts_with("AZpID.")), .funs = list(~ifelse(!is.na(.), paste(unlist(lapply(strsplit(x = ., split = ".", fixed = T), head, 1)), AZpID, sep = "."), NA)))
if (file.exists(paste(output.directory, "proteins_in_tissues.xlsx", sep = ""))) {
  file.remove(paste(output.directory, "proteins_in_tissues.xlsx", sep = ""))
}
write.to.excel(df_ = dia$index$all, file_ = paste(output.directory, "proteins_in_tissues.xlsx", sep = ""), sheet_ = "Sheet1", append_ = F)
dia$index$AZpID <- update.index(dia.index.all_ = dia$index$all %>% select(AZpID, Gene, Protein_Accession, Protein_Description, starts_with("AZpID.")), dia.filtered_ = dia$raw.filtered , tissues_ = tissues)

# average duplicate rotein accessions in tissue datasets
dia$raw.filtered <- lapply(dia$raw.filtered, average.duplicate.protein.accessions, dpa_ = tmp.azpid$duplicated)

# store other variables/results
dia$other <- list()

# make data frames proteins x samples and samples x proteins
dia$raw.filtered.unique.proteins <- merge.unique.proteins.dia(dia.all_ = dia$raw.filtered)
dia$raw.filtered.unique.samples <- merge.unique.samples.dia(dia.all_ = dia$raw.filtered)

dia$other$raw.unique.proteins <- merge.unique.proteins.dia(dia.all_ = dia$raw)
dia$other$raw.unique.samples <- merge.unique.samples.dia(dia.all_ = dia$raw)

# cleanup
rm(read.dia, get.dia.index, prepare.dia, unique.protein.accession, filter.proteins, filter.samples, update.index, average.duplicate.protein.accessions, dia.tmp, tmp.azpid)
```

## Exploring identified proteins in tissues and phenotypes

The table below shows the number of proteins identified in each tissue, the total number of identified proteins and the unique number of identified proteins before and after filtering. The same is shown for samples, including the dustribution of samples over phenotypes.

```{r show.table.protein.stats, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
# calculate distribution of samples over phenotypes
sample.distribution.in.phenos <- plyr::join_all(dfs = lapply(seq_along(dia$raw.filtered), function(i) covars %>% filter(id %in% tail(colnames(dia$raw.filtered[[i]]), -1)) %>% group_by(pheno_3c) %>% summarise(!!names(dia$raw.filtered)[i] := n())),
                                                by = 'pheno_3c', type = 'full') %>%
  as_tibble() %>%
  mutate(pheno_3c = replace(pheno_3c, is.na(pheno_3c), "NA"),
         Total = rowSums(.[2:6]),
         Unique = qlcMatrix::rowMax(.[2:6] %>% as.matrix()) %>% as.numeric()) %>%
  column_to_rownames("pheno_3c") %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Tissue")

# prepare data frame
number.proteins.per.tissue <- data.frame(Tissue = c(tissues, "Total", "Unique"),
                                         Samples.Raw = c(unname(unlist(lapply(dia$raw, ncol))) - 1, ncol(dia$other$raw.unique.proteins) - 1, nrow(dia$other$raw.unique.samples)),
                                         Proteins.Raw = c(unname(unlist(lapply(dia$raw, nrow))), ncol(dia$other$raw.unique.samples) - 1, nrow(dia$other$raw.unique.proteins)),
                                         Missing.Raw = c(unlist(lapply(dia$raw, function(i) paste(round(x = ((sum(is.na(i %>% column_to_rownames("AZpID"))) * 100) / (ncol(i %>% column_to_rownames("AZpID")) * nrow(i %>% column_to_rownames("AZpID")))), digits = 2), "%", sep = ""))), paste(round(x = (sum(unlist(lapply(dia$raw, function(i) sum(is.na(i %>% column_to_rownames("AZpID")))))) * 100) / sum(unlist(lapply(dia$raw, function(i) ncol(i %>% column_to_rownames("AZpID")) * nrow(i %>% column_to_rownames("AZpID"))))), digits = 2), "%", sep = ""), ""),
                                         Samples.Filtered = c(unname(unlist(lapply(dia$raw.filtered, ncol))) - 1, ncol(dia$raw.filtered.unique.proteins) - 1, nrow(dia$raw.filtered.unique.samples)),
                                         Proteins.Filtered = c(unname(unlist(lapply(dia$raw.filtered, nrow))), ncol(dia$raw.filtered.unique.samples) - 1, nrow(dia$raw.filtered.unique.proteins)),
                                         Missing.Filtered = c(unlist(lapply(dia$raw.filtered, function(i) paste(round(x = ((sum(is.na(i %>% column_to_rownames("AZpID"))) * 100) / (ncol(i %>% column_to_rownames("AZpID")) * nrow(i %>% column_to_rownames("AZpID")))), digits = 2), "%", sep = ""))), paste(round(x = (sum(unlist(lapply(dia$raw.filtered, function(i) sum(is.na(i %>% column_to_rownames("AZpID")))))) * 100) / sum(unlist(lapply(dia$raw, function(i) ncol(i %>% column_to_rownames("AZpID")) * nrow(i %>% column_to_rownames("AZpID"))))), digits = 2), "%", sep = ""), ""),
                                         Details.Samples.Removed = c(unlist(lapply(seq_along(tissues), function(i) paste(filter(covars, id %in% colnames(dia$raw[[i]])[!(colnames(dia$raw[[i]]) %in% colnames(dia$raw.filtered[[i]]))]) %>% mutate(pheno_3c = replace(pheno_3c, is.na(pheno_3c), "NA"), tmp = paste(id, " (", pheno_3c, ")", sep = "")) %>% pull(tmp), collapse = ", "))), "", paste(data.frame(s = unlist(lapply(seq_along(tissues), function(i) colnames(dia$raw[[i]])[!(colnames(dia$raw[[i]]) %in% colnames(dia$raw.filtered[[i]]))])), stringsAsFactors = F) %>% group_by(s) %>% summarise(n = n()) %>% mutate(p = paste(s, " (", n, ")", sep = "")) %>% pull(p), collapse = ", ")),
                                         stringsAsFactors = F) %>%
  mutate(Samples.Removed = Samples.Raw - Samples.Filtered,
         Percentage.Samples.Removed = paste(round((Samples.Removed * 100) / Samples.Raw, digits = 2), "%", sep = ""),
         Proteins.Removed = Proteins.Raw - Proteins.Filtered,
         Percentage.Proteins.Removed = paste(round((Proteins.Removed * 100) / Proteins.Raw, digits = 2), "%", sep = "")) %>%
  select(Tissue, ends_with("Raw"), ends_with("Filtered"), Samples.Removed, Percentage.Samples.Removed, Details.Samples.Removed, Proteins.Removed, Percentage.Proteins.Removed) %>%
  left_join(sample.distribution.in.phenos, by = 'Tissue')

if (file.exists(paste(output.directory, "number_of_proteins_per_tissue.xlsx", sep = ""))) {
  file.remove(paste(output.directory, "number_of_proteins_per_tissue.xlsx", sep = ""))
}
write.to.excel(df_ = number.proteins.per.tissue, file_ = paste(output.directory, "number_of_proteins_per_tissue.xlsx", sep = ""), sheet_ = "Sheet1", append_ = F)

kable(number.proteins.per.tissue, col.names = unlist(lapply(strsplit(x = colnames(number.proteins.per.tissue), split = ".", fixed = T), head, 1)), align = 'r') %>% 
  kable_styling(full_width = F, bootstrap_options = "striped") %>%
  column_spec(1, bold = T, border_right = T) %>%
  row_spec(6:7, bold = T) %>%
  add_header_above(c(" " = 1, "Raw" = 3, "Filtered" = 3, "Removed Samples" = 3, "Removed Proteins" = 2, "Phenotypes" = 4))

dia$other$raw.filtered.unique.proteins <- dia$raw.filtered.unique.proteins
dia$raw.filtered.unique.proteins <- NULL
dia$other$raw.filtered.unique.samples <- dia$raw.filtered.unique.samples
dia$raw.filtered.unique.samples <- NULL

rm(number.proteins.per.tissue)
```

The plot below shows the distributon of samples over phenotypes for each tissue in a doughnut plot. The inner doughnut represents samples in _CTRL_, _PD_ and _T2D_. The next outer doughnut represents the merged class of _CTRL_ and _PD_, while the outmost one the merged class of _PD_ and _T2D_.

```{r show.plots.protein.stats, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=9, fig.height=7}
sample.distribution.in.phenos %>%
  filter(Tissue %in% tissues) %>%
  select(Tissue, all_of(c("CTRL", "PD", "T2D"))) %>%
  mutate(Total = rowSums(.[2:4]),
         CTRL_PD = CTRL + PD,
         CTRL_PD__T2D = T2D,
         PD_T2D = PD + T2D,
         PD_T2D__CTRL = CTRL) %>%
  select(Tissue, CTRL, PD, T2D, CTRL_PD, CTRL_PD__T2D, PD_T2D, PD_T2D__CTRL, Total) %>%
  mutate_at(.vars = vars(CTRL:PD_T2D__CTRL), .funs = list(~((. * 100)/Total))) %>%
  select(-Total) %>%
  gather("Pheno", "Percentage", -Tissue) %>%
  mutate(Resolution = case_when(
    Pheno %in% c("CTRL", "PD", "T2D") ~ "3C",
    Pheno %in% c("CTRL_PD", "CTRL_PD__T2D") ~ "2C2",
    Pheno %in% c("PD_T2D", "PD_T2D__CTRL") ~ "2C1",
    TRUE ~ "TMP"),
    Resolution = replace(Resolution, Resolution == "TMP", NA)) %>%
  mutate_all(as.character) %>%
  map_df(~parse_guess(.)) %>%
  select(Tissue, Resolution, Pheno, Percentage) %>%
  arrange(Tissue, desc(Resolution), Pheno, Percentage) %>%
  mutate(Pheno = factor(x = Pheno, levels = c("CTRL", "PD", "T2D", "CTRL_PD", "CTRL_PD__T2D", "PD_T2D__CTRL", "PD_T2D")),
         Resolution = factor(x = Resolution, levels = c("3C", "2C2", "2C1")),
         Tissue = factor(x = Tissue, levels = tissues),
         Text.Percentage = paste(as.character(round(x = Percentage, digits = 0)), "%", sep = "")) %>%
  ggplot(aes(x = Resolution, y = Percentage, fill = Pheno)) +
  geom_bar(stat = "identity", color = "white", width = 0.7) +
  scale_x_discrete(limits = c(" ", "3C", "2C2", "2C1")) +
  geom_text(aes(label = Text.Percentage), position = position_stack(vjust = 0.5)) +
  scale_fill_manual(values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) +
  coord_polar(theta = "y", start = 0) +
  facet_wrap(~ Tissue) +
  theme_void()

rm(sample.distribution.in.phenos)
```

# Quality control (QC), normalization and imputation of the data

## QC of the data

The figure below shows the log~2~ FC of the median sample intensity over the average intensity of the medians of all samples. The left column shows the raw data (raw) and the right one for the fileterd set of median samples intensities deviating less than log~2~5 (filtered).

```{r plot.median.diff, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=12, fig.height=16}
plot.median.diff <- function(dia_, tissue_, pheno.col_) {
  return(select(dia_, -AZpID) %>%
           summarise_all(.funs = list(~median(x = ., na.rm = TRUE))) %>%
           t() %>%
           as.data.frame() %>%
           rename("median.intensity" = "V1") %>%
           rownames_to_column("id") %>%
           mutate(intensity.diff = abs(median.intensity) - mean(median.intensity),
                  id = as.character(id)) %>%
           left_join(select(covars, id, all_of(pheno.col_)) %>% rename("pheno" = all_of(pheno.col_)), by = 'id') %>%
           filter(!is.na(pheno)) %>%
           arrange(intensity.diff) %>%
           mutate(id = factor(id, levels = id)) %>%
           ggplot(aes(x = id, y = intensity.diff, color = pheno, fill = pheno)) +
           geom_bar(stat="identity") +
           theme_minimal() + 
           # geom_hline(yintercept=c(-log2(5), log2(5)), linetype="dashed", color = "red") +
           scale_y_continuous(breaks = seq(from = -5, to = 5, by = 1), labels = seq(from = -5, to = 5, by = 1), limits = c(-5, 5),
                              sec.axis = sec_axis(~sign(.)*(2^abs(.)), breaks = c(-25, -10, -5, 0, 5, 10, 25), labels = c(-25, -10, -5, 0, 5, 10, 25), name = "linear scale")) +
           labs(title = tissue_, y = expression('log'[2]*'(sample'['avg']*') - log'[2]*'(total'['avg']*')')) +
           scale_fill_manual(name = "Phenotype", values=setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
           scale_color_manual(name = "Phenotype", values=setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
           theme(axis.ticks.x = element_blank(),
                 axis.text.x = element_text(angle = 90),
                 axis.title.x = element_blank(),
                 panel.grid.minor = element_blank(),
                 panel.grid.major.x = element_blank(),
                 plot.title = element_text(hjust = 0.5)))
}

plot_grid(plot_grid(plotlist = lapply(seq_along(dia$raw), function(i) plot.median.diff(dia_ = dia$raw[[i]], tissue_ = names(dia$raw)[i], pheno.col_ = covars.pheno)), ncol = 1),
          plot_grid(plotlist = lapply(seq_along(dia$raw.filtered), function(i) plot.median.diff(dia_ = dia$raw.filtered[[i]], tissue_ = names(dia$raw.filtered)[i], pheno.col_ = covars.pheno)), ncol = 1), ncol = 2,
          labels = c("Raw", "Filtered"))

rm(plot.median.diff)
```

## Missingness of the data

The plots below show the characteristic of the missing data from the dataset. The left panels show the intensity distributions of the proteins that have missing values, and the right panels show the cumulative fraction of proteins with and without missing values. Every row represents a tissues.

```{r explore.missingness, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=6, fig.height=10}
plot.miss <- function(dia_, tissue_) {
  dia.tmp_ <- dia_ %>%
    mutate(total.tmp = select(., -AZpID) %>% rowSums(., na.rm = T),
           count = select(., -AZpID) %>% is.na(.) %>% `!` %>% rowSums(., na.rm = T),
           total = length(colnames(.)) - 1,
           mean = total.tmp / count,
           miss = ifelse(count < total, TRUE, FALSE)) %>%
    select(mean, miss) %>%
    group_by(miss) %>%
    arrange(mean) %>%
    mutate(num = 1, cs = cumsum(num), cs_frac = cs/n()) %>%
    ungroup()
  
  # Plot the densities and cumalitive fractions for
  # proteins with and without missing values
  p1 <- ggplot(dia.tmp_, aes(mean, col = miss)) +
    geom_density(na.rm = TRUE) +
    xlim(10, 30) +
    ylim(0, 0.5) +
    labs(x = expression(log[2]~"Intensity"), y = "Density", title = tissue_) +
    guides(col = guide_legend(title = "Missing values")) +
    theme_minimal() + 
    theme(panel.grid.minor = element_blank())
  
  p2 <- ggplot(dia.tmp_, aes(mean, cs_frac, col = miss)) +
    geom_line() +
    xlim(10, 30) +
    ylim(0, 1) +
    labs(x = expression(log[2]~"Intensity"), y = "Cumulative fraction", title = tissue_) +
    guides(col = guide_legend(title = "Missing values")) + 
    theme_minimal() + 
    theme(panel.grid.minor = element_blank())
  
  return(plot_grid(p1, p2, nrow = 1, labels = c(tissue_, "")))
}

plot_grid(plotlist = lapply(tissues, function(i) plot.miss(dia_ = dia$raw.filtered[[i]], tissue_ = names(dia$raw.filtered)[i])), ncol = 1)

rm(plot.miss)
```

## Normalization of the data

```{r normalizing.the.data, warning=FALSE, message=FALSE, echo=FALSE, results='hide'}
normalize.median <- function(dia_) {
  return(mutate_at(dia_, .vars = vars(-AZpID), .funs = list(~(. - median(., na.rm = T)))))
}

dia$normalized <- lapply(dia$raw.filtered, normalize.median)
dia$other$normalized.unique.proteins <- merge.unique.proteins.dia(dia.all_ = dia$normalized)
dia$other$normalized.unique.samples <- merge.unique.samples.dia(dia.all_ = dia$normalized)

rm(normalize.median)
```

The figure below shows the raw (left) and normalized protein intensities (right). Normalization is applied to remove systematic differences in protein abundance due to different sample concentrations, or different amounts of sample loaded in the columns. Normalization is important, so that true differentially expressed proteins can be detected. To do this the z-score of the log~2~ transformed intensities is computed, which is updated by the average of the standard deviation of the log~2~ transformed intensities in all samples. After normalization all samples have a similar distribution.

```{r explore.normalization, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=16, fig.height=35}
plot.intensity.distributions <- function(dia_, pheno.col_, tissue_, ylim.min_, ylim.max_) {
  return(dia_ %>%
           gather("variable", "value", -AZpID) %>%
           left_join(select(covars, id, all_of(pheno.col_)) %>% rename("pheno" = all_of(pheno.col_)), by = c("variable" = "id")) %>%
           filter(!is.na(pheno)) %>%
           arrange(match(pheno, sort(unique(pheno))), variable) %>%
           mutate(pheno = factor(x = pheno, levels = c(sort(unique(pheno)), "NA")),
                  variable = factor(x = variable, levels = unique(variable))) %>%
           ggplot(aes(x = variable, y = value, color = pheno, fill = pheno)) +
           labs(title = tissue_, y = expression('log'[2]*' intensity')) +
           geom_boxplot(alpha = 0.5) +
           theme_minimal() +
           coord_flip() +
           ylim(ylim.min_, ylim.max_) +
           scale_fill_manual(name = "Phenotype", values=setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
           scale_color_manual(name = "Phenotype", values=setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
           theme(axis.ticks.y = element_blank(),
                 axis.title.y = element_blank(),
                 plot.title = element_text(hjust = 0.5)))
}

plot_grid(plot_grid(plotlist = lapply(tissues, function(i) plot.intensity.distributions(dia_ = dia$raw.filtered[[i]], pheno.col_ = covars.pheno, tissue_ = i, ylim.min_ = plyr::round_any(x = min(dia$other$raw.filtered.unique.proteins %>% select(-AZpID), na.rm = T), accuracy = 5, f = ceiling), ylim.max_ = plyr::round_any(x = max(dia$other$raw.filtered.unique.proteins %>% select(-AZpID), na.rm = T), accuracy = 5, f = ceiling))), ncol = 1),
          plot_grid(plotlist = lapply(tissues, function(i) plot.intensity.distributions(dia_ = dia$normalized[[i]], pheno.col_ = covars.pheno, tissue_ = i, ylim.min_ = min(unlist(lapply(dia$normalized, function(i) plyr::round_any(x = min(i %>% select(-AZpID), na.rm = T), accuracy = 5, f = ceiling)))), ylim.max_ = max(unlist(lapply(dia$normalized, function(i) plyr::round_any(x = max(i %>% select(-AZpID), na.rm = T), accuracy = 5, f = ceiling)))))), ncol = 1),
          ncol = 2, labels = c("Raw", "Normalized"))

rm(plot.intensity.distributions)
```

## Imputation of the data

The density plots below summarize the imputed data using QRILC and compare it to the non-imputed normalized set (top row). Every column repsesents one tissue and every colored line represents one sample.

```{r explore.imputation, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=15, fig.height=9}
impute.and.plot <- function(dia_, tissue_) {
  dia.imputed.QRILC <- imputeLCMD::impute.QRILC(dataSet.mvs = dia_ %>% column_to_rownames("AZpID"))
  
  p1 <- dia_ %>%
    gather("sample", "value", -AZpID) %>%
    ggplot(data = ., aes(x = value, color = sample)) + 
    geom_density(na.rm = TRUE) +
    labs(title = tissue_, y = ifelse(tissue_ == tissues[1], "Normalized", "")) +
    xlim(-15, 15) +
    ylim(0, 0.25) +
    theme_minimal() + 
    theme(legend.position = "none",
          axis.title.x = element_blank(),
          panel.grid.minor = element_blank())
  
  p2 <- dia.imputed.QRILC[[1]] %>%
    rownames_to_column("AZpID") %>%
    gather("sample", "value", -AZpID) %>%
    ggplot(data = ., aes(x=value, color = sample)) + 
    geom_density() +
    labs(y = ifelse(tissue_ == tissues[1], "QRILC Imputed", "")) +
    xlim(-15, 15) +
    ylim(0, 0.25) +
    theme_minimal() + 
    theme(legend.position = "none",
          axis.title.x = element_blank(),
          panel.grid.minor = element_blank())
  
  p <- plot_grid(p1, p2, ncol = 1)
  
  return(list(QRILC = dia.imputed.QRILC[[1]] %>% rownames_to_column("AZpID"), plot = p))
}

imputation.result <- lapply(seq_along(tissues), function(i) impute.and.plot(dia_ = dia$normalized[[i]], tissue_ = names(dia$normalized)[i]))

dia$normalized.imputed <- lapply(imputation.result, function(i) i[["QRILC"]])
names(dia$normalized.imputed) <- tissues
dia$normalized.imputed.unique.proteins <- merge.unique.proteins.dia(dia.all_ = dia$normalized.imputed)
dia$normalized.imputed.unique.samples <- merge.unique.samples.dia(dia.all_ = dia$normalized.imputed)

if(file.exists(paste(output.directory, "normalized_imputed_dataset.xlsx", sep = ""))) {
  file.remove(paste(output.directory, "normalized_imputed_dataset.xlsx", sep = ""))
}
lapply(seq_along(tissues), function(i) write.to.excel(df_ = dia$normalized.imputed[[i]] %>% left_join(dia$index$index, by = 'AZpID'), file_ = paste(output.directory, "normalized_imputed_dataset.xlsx", sep = ""), sheet_ = names(dia$normalized.imputed)[i], append_ = ifelse(i == 1, FALSE, TRUE)))

plot_grid(plotlist = lapply(imputation.result, function(i) i[["plot"]]), ncol = length(tissues))

rm(impute.and.plot, imputation.result, merge.unique.proteins.dia, merge.unique.samples.dia, merge.full.joins)
```

# Visualizing the data and the results

## Shared proteins across tissues

The figure shows shared proteins among tissues. The left panel (A) shows shared proteins among all tissues, while the right one (B) among those with more than 1000 identified proteins.

```{r venn.diagram.shared.proteins, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.ext='png', fig.width=14, fig.height=7, dev='png'}
suppressPackageStartupMessages(library(VennDiagram))

draw.venn.diagram.5.tissues <- function(index.azpid_, filename_) {
  venn.diagram(x = list(Fat = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[1], sep = ".")))) %>% pull(AZpID),
                        Liver = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[2], sep = ".")))) %>% pull(AZpID),
                        Muscle = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[3], sep = ".")))) %>% pull(AZpID),
                        Pancreas = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[4], sep = ".")))) %>% pull(AZpID),
                        Serum = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[5], sep = "."))))%>% pull(AZpID)),
               category.names = c(paste("VAT", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[1], sep = "."))))), ")", sep = ""),
                                  paste("Liver", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[2], sep = "."))))), ")", sep = ""),
                                  paste("Muscle", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[3], sep = "."))))), ")", sep = ""),
                                  paste("Islets", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[4], sep = "."))))), ")", sep = ""),
                                  paste("Serum", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[5], sep = "."))))), ")", sep = "")),
               filename = filename_, imagetype = "tiff", lwd = 1, cex = 0.7, fontfamily = "sans",
               col = venn.cols, fill = c(alpha(colour = venn.cols[1], alpha = 0.3), alpha(colour = venn.cols[2], alpha = 0.3), alpha(colour = venn.cols[3], alpha = 0.3),
                                         alpha(colour = venn.cols[4], alpha = 0.3), alpha(colour = venn.cols[5], alpha = 0.3)),
               cat.cex = 1.5, cat.default.pos = "outer", cat.pos = c(-65, 163, -100, 130, 20), cat.dist = c(0.27, -0.24, 0.25, 0.22, 0.23), cat.fontfamily = "sans", cat.col = venn.cols)
}

draw.venn.diagram.4.tissues <- function(index.azpid_, filename_) {
  venn.diagram(x = list(Fat = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[1], sep = ".")))) %>% pull(AZpID),
                        Liver = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[2], sep = ".")))) %>% pull(AZpID),
                        Muscle = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[3], sep = ".")))) %>% pull(AZpID),
                        Pancreas = filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[4], sep = ".")))) %>% pull(AZpID)),
               category.names = c(paste("VAT", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[1], sep = "."))))), ")", sep = ""),
                                  paste("Liver", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[2], sep = "."))))), ")", sep = ""),
                                  paste("Muscle", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[3], sep = "."))))), ")", sep = ""),
                                  paste("Islets", "\n(n=", nrow(filter(index.azpid_, !is.na(!!as.symbol(paste("AZpID", tissues[4], sep = "."))))), ")", sep = "")),
               filename = filename_, imagetype = "tiff", lwd = 1, cex = 2, fontfamily = "sans", label.col = 'grey30',
               col = venn.cols[1:4], fill = c(alpha(colour = venn.cols[1], alpha = 0.3), alpha(colour = venn.cols[2], alpha = 0.3), alpha(colour = venn.cols[3], alpha = 0.3), alpha(colour = venn.cols[4], alpha = 0.3)),
               cat.cex = 1.5, cat.default.pos = "outer", cat.pos = c(-13, 193, -13, 193), cat.dist = c(0.24, -0.24, 0.14, -0.14), cat.fontfamily = "sans", cat.col = venn.cols[1:4])
}

venn.tiff.files <- c(paste(output.directory, "venn.diagram.shared.proteins.5.tissues-1.tiff", sep = ""), paste(output.directory, "venn.diagram.shared.proteins.4.tissues-1.tiff", sep = ""))
venn.cols <- unname(tissue.colors)
draw.venn.diagram.5.tissues(index.azpid_ = dia$index$AZpID, filename_ = venn.tiff.files[1])
draw.venn.diagram.4.tissues(index.azpid_ = dia$index$AZpID, filename_ = venn.tiff.files[2])

plot_grid(ggdraw() + draw_image(venn.tiff.files[1]), ggdraw() + draw_image(venn.tiff.files[2]), labels = c("A", "B"), label_size = 20, ncol = 2, scale = 0.95)

file.remove(paste(output.directory, list.files(path = output.directory, pattern = glob2rx("venn*.log")), sep = ""))
rm(venn.tiff.files, venn.cols, draw.venn.diagram.4.tissues, draw.venn.diagram.5.tissues)
```

## Clustering the proteomics into tissues

The figure below shows the hierarchical clustering (HC) for normalized and imputed protein levels across tissues. Proteins are clustered based on their _1 - r_, where _r_ is the Pearson correlation coefficient across protein abundancies. The selected HC method is _ward.D2_. The first row below the dendrogram indicates tissues and the next one indicates T2D phenotypes. The barplot in the bottom shows the number of identified proteins in each sample prior to filtering and the black line indicates the number of proteins after filtering.

```{r cluster.tissues.for.proteins, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=14, fig.height=10}
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(circlize))

na.samples <- expand.grid(tissues, (covars %>% filter(is.na(pheno_3c)) %>% pull(id)), stringsAsFactors = F) %>%
  mutate(Var3 = paste(Var1, Var2, sep = ".")) %>%
  pull(Var3)
dia.norm.unique.proteins.matrix <- dia$normalized.imputed.unique.proteins %>%
  select(-any_of(na.samples)) %>%
  column_to_rownames("AZpID") %>%
  as.matrix()
# distance matrix from 1-Pearson correlation
heatmap.distance.matrix <- 1 - cor(x = dia.norm.unique.proteins.matrix, use = "pairwise.complete.obs", method = "pearson")
# hierarchical clustering of the distance matrix
hclust.heatmap <- hclust(as.dist(heatmap.distance.matrix), method = "complete")
# annotation data frame for annotation bars at the top of te heatmap
heatmap.annotation.df <- data.frame(sample.name = colnames(dia.norm.unique.proteins.matrix), stringsAsFactors = F) %>%
  mutate(tissue = unlist(lapply(strsplit(x = sample.name, split = ".", fixed = T), head, 1))) %>%
  left_join(as.data.frame(tissue.colors) %>% rownames_to_column("tissue"), by = 'tissue') %>%
  mutate_if(is.factor, as.character) %>%
  mutate(plain.sample.name = unlist(lapply(strsplit(x = sample.name, split = ".", fixed = T), tail, 1))) %>%
  left_join(covars %>% select(id, pheno_3c), by = c('plain.sample.name' = 'id'))
# Create the heatmap annotation
heatmap.tissue.annotation <- HeatmapAnnotation(df = heatmap.annotation.df %>% select(tissue, pheno_3c),
                                               col = list(tissue = tissue.colors,
                                                          pheno_3c = setNames(object = filter(phenotypes.colors, class == "pheno_3c") %>% pull(colors), nm = filter(phenotypes.colors, class == "pheno_3c") %>% pull(names))),
                                               show_annotation_name = FALSE)
# Prepare cloring of the heatmap according to tissue.colors
dendrogram.vars <- data.frame(colors = mutate(heatmap.annotation.df, row.num = row_number()) %>% arrange(match(row.num, hclust.heatmap$order)) %>% pull(tissue.colors) %>% unique(),
                              names = mutate(heatmap.annotation.df, row.num = row_number()) %>% arrange(match(row.num, hclust.heatmap$order)) %>% pull(tissue) %>% unique(),
                              stringsAsFactors = F) %>%
  mutate(nrow = row_number()) %>%
  rowwise() %>%
  mutate(num.imputed.proteins = nrow(dia$index$AZpID %>% select(all_of(paste("AZpID", names, sep = "."))) %>% filter(complete.cases(.)))) %>%
  ungroup()

dia.norm.nonimputed.unique.proteins.matrix <- dia$other$raw.unique.proteins %>%
  select(AZpID, all_of(heatmap.annotation.df$sample.name)) %>%
  mutate_at(.vars = vars(-AZpID), .funs = list(~(ifelse(is.na(.), 0, 1)))) %>%
  column_to_rownames("AZpID") %>%
  as.matrix() %>%
  colSums()

heatmap.barplot.annotation = HeatmapAnnotation(
  id.proteins = anno_barplot(
    dia.norm.nonimputed.unique.proteins.matrix, 
    bar_width = 1, 
    gp = gpar(col = "white", fill = "#ffe200ff"), 
    border = FALSE,
    axis_param = list(at = seq(from = 0, to = 5000, by = 2500),
                      labels = as.character(seq(from = 0, to = 5000, by = 2500))),
    height = unit(4, "cm"), 
  ), show_annotation_name = FALSE)

if (plyr::round_any(x = min(dia.norm.unique.proteins.matrix, na.rm = T), accuracy = 10, f = floor) == 0) {
  heatmap.colors.gradient <- colorRamp2(breaks = seq(from = plyr::round_any(x = min(dia.norm.unique.proteins.matrix, na.rm = T), accuracy = 10, f = floor),
                                                     to = plyr::round_any(x = max(dia.norm.unique.proteins.matrix, na.rm = T), accuracy = 10, f = ceiling),
                                                     by = 5),
                                        colors = c('#f7efe2ff', '#fef2e4ff', '#fd974fff', '#c60000ff', '#805a3bff'),
                                        transparency = 0, space = "RGB")
} else {
  heatmap.colors.gradient <- colorRamp2(breaks = seq(from = plyr::round_any(x = min(dia.norm.unique.proteins.matrix, na.rm = T), accuracy = 10, f = floor),
                                                     to = plyr::round_any(x = max(dia.norm.unique.proteins.matrix, na.rm = T), accuracy = 10, f = ceiling),
                                                     by = 10),
                                        colors = c('#192e5bff', '#1d65a6ff', '#f7efe2ff', '#c60000ff', '#805a3bff'),
                                        transparency = 0, space = "RGB")
}

barplot.line.color <- '#333333ff'
barplot.line.thickness <- unit(1, "cm")

Heatmap(matrix = dia.norm.unique.proteins.matrix,
        name = "protein level",
        na_col = '#fcfcfcff', #grey90
        col = heatmap.colors.gradient,
        show_row_names = FALSE,
        cluster_rows = FALSE,
        #cluster_columns = dendextend::color_branches(hclust.heatmap, k = 5, col = dendrogram.vars$colors),
        clustering_distance_columns = "pearson", # not working as expected
        clustering_method_columns = "ward.D2", # goes with the one above
        column_dend_side = "top",
        column_dend_height = unit(4, "cm"),
        column_split = 5,
        column_title_gp = gpar(fontsize = 16, fontface = "bold", col = "grey40"),
        show_column_names = FALSE,
        top_annotation = heatmap.tissue.annotation,
        bottom_annotation = heatmap.barplot.annotation)

decorate_annotation("id.proteins", {
  grid.lines(x = unit(c(dendrogram.vars$nrow[1] - 1, dendrogram.vars$nrow[1]), "npc"),
             y = unit(c(dendrogram.vars$num.imputed.proteins[1], dendrogram.vars$num.imputed.proteins[1]), "native"),
             gp = gpar(lty = barplot.line.thickness, col = barplot.line.color))
  grid.lines(x = unit(c(dendrogram.vars$nrow[1], dendrogram.vars$nrow[2]), "npc"),
             y = unit(c(dendrogram.vars$num.imputed.proteins[2], dendrogram.vars$num.imputed.proteins[2]), "native"),
             gp = gpar(lty = barplot.line.thickness, col = barplot.line.color))
  grid.lines(x = unit(c(dendrogram.vars$nrow[2], dendrogram.vars$nrow[3]), "npc"),
             y = unit(c(dendrogram.vars$num.imputed.proteins[3], dendrogram.vars$num.imputed.proteins[3]), "native"),
             gp = gpar(lty = barplot.line.thickness, col = barplot.line.color))
  grid.lines(x = unit(c(dendrogram.vars$nrow[3], dendrogram.vars$nrow[4]), "npc"),
             y = unit(c(dendrogram.vars$num.imputed.proteins[4], dendrogram.vars$num.imputed.proteins[4]), "native"),
             gp = gpar(lty = barplot.line.thickness, col = barplot.line.color))
  grid.lines(x = unit(c(dendrogram.vars$nrow[4], dendrogram.vars$nrow[5]), "npc"),
             y = unit(c(dendrogram.vars$num.imputed.proteins[5], dendrogram.vars$num.imputed.proteins[5]), "native"),
             gp = gpar(lty = barplot.line.thickness, col = barplot.line.color))
})

rm(dia.norm.unique.proteins.matrix, heatmap.distance.matrix, hclust.heatmap, heatmap.annotation.df, heatmap.tissue.annotation, dendrogram.vars, dia.norm.nonimputed.unique.proteins.matrix, heatmap.barplot.annotation, heatmap.colors.gradient, barplot.line.color, barplot.line.thickness, na.samples)
```

## Exploring the data in low dimensionality

### Exploring the data in the principal component analysis (PCA) space

The figure below shows a collection of 2D PCA plots for the 2 first principal components (PCs) for every tissue and combination of phenotypes. We use it to explore potential separation of groups in the PCA space. The first column of plots shows the three phenotypic classes _CTRL_, _PD_ and _T2D_, the second one showes the merged classes of _CTRL+PD_ and _T2D_, and the third one the merged classes of _PD+T2D_ and _CTRL_.

```{r pca.exploration, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=20, fig.height=16}
pca.plots <- function(dia_, pheno.col_, tissue_) {
  tmp.covars_ <- select(covars, id, all_of(pheno.col_)) %>%
    rename_at(.vars = vars(starts_with("pheno")), .funs = list(~make.unique(unlist(lapply(strsplit(x = ., split = "_", fixed = T), head, 1))))) %>%
    filter(id %in% tail(colnames(dia_), -1)) %>%
    filter(complete.cases(.)) %>%
    select(id, everything()) %>%
    as.data.frame()
  
  dia_ <- dia_ %>%
    select(AZpID, all_of(tmp.covars_$id)) %>%
    column_to_rownames("AZpID") %>%
    t()
  
  # pcaMethods::listPcaMethods
  # "svd", "nipals", "rnipals", "bpca", "ppca", "svdImpute", "robustPca", "nlpca", "llsImpute", "llsImputeAll"
  dia.pca_ <- pcaMethods::pca(object = dia_, nPcs = 2, method = "ppca", scale = "none", center = FALSE, completeObs = FALSE) # run pca
  dia.pca.scores_ <- as.data.frame(pcaMethods::scores(dia.pca_))
  
  p1 <- ggplot(dia.pca.scores_, aes(PC1, PC2, color = tmp.covars_[,2], label = tmp.covars_$id)) +
    geom_text() +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  p2 <- ggplot(dia.pca.scores_, aes(PC1, PC2, color = tmp.covars_[,3], label = tmp.covars_$id)) +
    geom_text() +
    labs(title = tissue_) +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  p3 <- ggplot(dia.pca.scores_, aes(PC1, PC2, color = tmp.covars_[,4], label = tmp.covars_$id)) +
    geom_text() +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  return(plot_grid(p1, p2, p3, ncol = 3))
}

plot_grid(plotlist = lapply(tissues, function(i) pca.plots(dia_ = dia$normalized.imputed[[i]], pheno.col_ = c("pheno_3c", "pheno_ctrl_pd", "pheno_pd_t2d"), tissue_ = i)), nrow = length(tissues))

rm(pca.plots)
```

### Exploring the data in the t-distributed stochastic neighbor embedding (tSNE) space

The figure below shows a collection of 2D tSNE plots for the 2 first embeddings for every tissue and combination of phenotypes. We use it to explore potential separation of groups in the tSNE space. The first column of plots shows the three phenotypic classes _CTRL_, _PD_ and _T2D_, the second one showes the merged classes of _CTRL+PD_ and _T2D_, and the third one the merged classes of _PD+T2D_ and _CTRL_.

```{r tsne.exploration, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=20, fig.height=16}
tsne.plots <- function(dia_, pheno.col_, tissue_) {
  tmp.covars_ <- select(covars, id, all_of(pheno.col_)) %>%
    rename_at(.vars = vars(starts_with("pheno")), .funs = list(~make.unique(unlist(lapply(strsplit(x = ., split = "_", fixed = T), head, 1))))) %>%
    filter(id %in% tail(colnames(dia_), -1)) %>%
    filter(complete.cases(.)) %>%
    select(id, everything()) %>%
    as.data.frame()
  
  dia_ <- dia_ %>% select(AZpID, all_of(tmp.covars_$id))
  dia.tmp_ <- data.table::transpose(l = dia_)
  colnames(dia.tmp_) <- dia.tmp_[1,]
  dia.tmp_ <- apply(dia.tmp_[2:nrow(dia.tmp_),], 2, as.numeric)
  rownames(dia.tmp_) <- tail(colnames(dia_), -1)
  dia_ <- dia.tmp_
  rm(dia.tmp_)
  
  # pcaMethods::listPcaMethods
  # "svd", "nipals", "rnipals", "bpca", "ppca", "svdImpute", "robustPca", "nlpca", "llsImpute", "llsImputeAll"
  dia.pca_ <- pcaMethods::pca(object = dia_, nPcs = 10, method = "ppca", scale = "none", center = FALSE, completeObs = FALSE) # run pca
  dia.pca.scores_ <- as.data.frame(pcaMethods::scores(dia.pca_))
  
  dia.tsne_ <- Rtsne::Rtsne(X = dia.pca.scores_, dims = 2, initial_dims = 50, perplexity = ((nrow(dia_) - 1) / 3), theta = 0.5, check_duplicates = TRUE,
                            pca = FALSE, partial_pca = FALSE, pca_center = FALSE, pca_scale = FALSE, normalize = FALSE)
  
  p1 <- ggplot(as.data.frame(dia.tsne_$Y), aes(V1, V2, color = tmp.covars_[,2], label = tmp.covars_$id)) +
    geom_text() +
    labs(x = "tSNE 1", y = "tSNE 2") +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  p2 <- ggplot(as.data.frame(dia.tsne_$Y), aes(V1, V2, color = tmp.covars_[,3], label = tmp.covars_$id)) +
    geom_text() +
    labs(title = tissue_, x = "tSNE 1", y = "tSNE 2") +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  p3 <- ggplot(as.data.frame(dia.tsne_$Y), aes(V1, V2, color = tmp.covars_[,4], label = tmp.covars_$id)) +
    geom_text() +
    labs(x = "tSNE 1", y = "tSNE 2") +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  return(plot_grid(p1, p2, p3, ncol = 3))
}

plot_grid(plotlist = lapply(tissues, function(i) tsne.plots(dia_ = dia$normalized.imputed[[i]], pheno.col_ = c("pheno_3c", "pheno_ctrl_pd", "pheno_pd_t2d"), tissue_ = i)), nrow = length(tissues))

rm(tsne.plots)
```

### Exploring the data in the uniform manifold approximation and projection (UMAP) space

The figure below shows a collection of 2D UMAP plots for the 2 first embeddings for every tissue and combination of phenotypes. We use it to explore potential separation of groups in the UMAP space. The first column of plots shows the three phenotypic classes _CTRL_, _PD_ and _T2D_, the second one showes the merged classes of _CTRL+PD_ and _T2D_, and the third one the merged classes of _PD+T2D_ and _CTRL_.

```{r umap.exploration, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=20, fig.height=16}
umap.plots <- function(dia_, pheno.col_, tissue_) {
  tmp.covars_ <- select(covars, id, all_of(pheno.col_)) %>%
    rename_at(.vars = vars(starts_with("pheno")), .funs = list(~make.unique(unlist(lapply(strsplit(x = ., split = "_", fixed = T), head, 1))))) %>%
    filter(id %in% tail(colnames(dia_), -1)) %>%
    filter(complete.cases(.)) %>%
    select(id, everything()) %>%
    as.data.frame()
  
  dia_ <- dia_ %>% select(AZpID, all_of(tmp.covars_$id))
  dia.tmp_ <- data.table::transpose(l = dia_)
  colnames(dia.tmp_) <- dia.tmp_[1,]
  dia.tmp_ <- apply(dia.tmp_[2:nrow(dia.tmp_),], 2, as.numeric)
  rownames(dia.tmp_) <- tail(colnames(dia_), -1)
  dia_ <- dia.tmp_
  rm(dia.tmp_)
  
  # pcaMethods::listPcaMethods
  # "svd", "nipals", "rnipals", "bpca", "ppca", "svdImpute", "robustPca", "nlpca", "llsImpute", "llsImputeAll"
  dia.pca_ <- pcaMethods::pca(object = dia_, nPcs = 10, method = "ppca", scale = "none", center = FALSE, completeObs = FALSE) # run pca
  dia.pca.scores_ <- as.data.frame(pcaMethods::scores(dia.pca_))
  
  custom.config <- umap::umap.defaults
  custom.config$n_neighbors <- 5
  dia.umap_ <- umap::umap(d = dia.pca.scores_, config = custom.config)
  
  p1 <- ggplot(as.data.frame(dia.umap_$layout), aes(V1, V2, color = tmp.covars_[,2], label = tmp.covars_$id)) +
    geom_text() +
    labs(x = "UMAP 1", y = "UMAP 2") +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  p2 <- ggplot(as.data.frame(dia.umap_$layout), aes(V1, V2, color = tmp.covars_[,3], label = tmp.covars_$id)) +
    geom_text() +
    labs(title = tissue_, x = "UMAP 1", y = "UMAP 2") +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  p3 <- ggplot(as.data.frame(dia.umap_$layout), aes(V1, V2, color = tmp.covars_[,4], label = tmp.covars_$id)) +
    geom_text() +
    labs(x = "UMAP 1", y = "UMAP 2") +
    theme_minimal() +
    stat_ellipse() +
    scale_color_manual(name = "Phenotype", values = setNames(object = phenotypes.colors$colors, nm = phenotypes.colors$names)) + 
    theme(plot.title = element_text(hjust = 0.5))
  
  return(plot_grid(p1, p2, p3, ncol = 3))
}

plot_grid(plotlist = lapply(tissues, function(i) umap.plots(dia_ = dia$normalized.imputed[[i]], pheno.col_ = c("pheno_3c", "pheno_ctrl_pd", "pheno_pd_t2d"), tissue_ = i)), nrow = length(tissues))

rm(umap.plots)
```

## Exploring the the impact of clinical and anthropometric variables

The plot below shows the canonical correlation analysis (CCA) of the set of clinical and anthropometric variables. The intensity of the correlation is explained in the legend.

```{r correlation.of.explanatory.variables, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=6, fig.height=6}
form <- ~ sex + age + blood.group + bmi + hba1c + gsis + islets.purity + cit.min + days.in.icu + antidiuretics + antiepileptic + antiinfectives + bba + bsps + cardiac.therapy + diuretics + glucocorticoids + insulin + pheno_3c + pheno_ctrl_pd + pheno_pd_t2d

# Compute Canonical Correlation Analysis (CCA) between all pairs of variables and return absolute correlation values
var.cor.mat = variancePartition::canCorPairs(form, covars)

# Plot correlation matrix
corrplot::corrplot(corr = var.cor.mat, order = "hclust", is.corr = F, hclust.method = "ward.D2", diag = F, type = "lower", tl.col = "black", tl.srt = 45)

rm(form, var.cor.mat)
```

The plot below shows the percentage of the variance of all proteins that is explained by each clinical and anthropometric variable. The table on the right hand side shows the details of the explained variance (median, mean and sd). The variables we choose to use later to account for on our models are those with more than 1% of median variance. Sex seems to be explaining very little variance overall.

```{r variance.by.explanatory.variables, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=10, fig.height=20}
run.varPart <- function(dia_, form_, snow.workers_, tissue_) {
  dia_ <- dia_ %>% column_to_rownames("AZpID") %>% filter(complete.cases(.)) # variancePartitionaccepts only non-NA rows
  tmp.covars_ <- covars %>% right_join(data.frame(id = colnames(dia_), stringsAsFactors = F), by = 'id') # arrange IDs of covariates according to dia_
  varPart <- variancePartition::fitExtractVarPartModel(exprObj = dia_ , formula = form_, data = tmp.covars_, BPPARAM = BiocParallel::SnowParam(workers = snow.workers_, type = "SOCK"))
  
  # sort variables (i.e. columns) by median fraction of variance explained
  vp <- variancePartition::sortCols(varPart)
  
  # violin plot of contribution of each variable to total variance
  p <- variancePartition::plotVarPart(obj = vp, main = tissue_)
  
  vp.tab <- bind_cols(as.list(vp)) %>%
    t() %>%
    as.data.frame()
  vp.tab$mean <- apply(vp.tab, 1, mean)
  vp.tab$median <- apply(vp.tab, 1, median)
  vp.tab$sd <- apply(vp.tab, 1, sd)
  vp.tab <- vp.tab %>%
    select(mean, median, sd) %>%
    rownames_to_column("Variable")
  
  return(list(plot = grid.arrange(p, tableGrob(vp.tab), nrow = 1), tab = vp.tab))
}

form <- ~ (1|sex) + (1|blood.group) + (1|antidiuretics) + (1|antiepileptic) + (1|antiinfectives) + (1|bba) + (1|bsps) + (1|cardiac.therapy) + (1|diuretics) + (1|glucocorticoids) + (1|insulin) + age + bmi + cit.min + days.in.icu
form.pancreas <- ~ (1|sex) + (1|blood.group) + (1|antidiuretics) + (1|antiinfectives) + (1|bba) + (1|bsps) + (1|cardiac.therapy) + (1|diuretics) + (1|glucocorticoids) + (1|insulin) + age + bmi + islets.purity + cit.min + days.in.icu
varpart.result <- lapply(seq_along(tissues),
                         function(i) if (names(dia$normalized.imputed)[i] == "Pancreas") {
                           run.varPart(dia_ = dia$normalized.imputed[[i]], form_ =  form.pancreas, snow.workers_ = num.of.cores.total, tissue_ = names(dia$normalized.imputed)[i])
                         } else {
                           run.varPart(dia_ = dia$normalized.imputed[[i]], form_ =  form, snow.workers_ = num.of.cores.total, tissue_ = names(dia$normalized.imputed)[i])
                         })
names(varpart.result) <- tissues

if(file.exists(paste(output.directory, "variance_by_explanatory_variables.xlsx", sep = ""))) {
  file.remove(paste(output.directory, "variance_by_explanatory_variables.xlsx", sep = ""))
}
lapply(seq_along(tissues), function(i) write.to.excel(df_ = varpart.result[[i]]$tab, file_ = paste(output.directory, "variance_by_explanatory_variables.xlsx", sep = ""), sheet_ = names(varpart.result)[i], append_ = ifelse(i == 1, FALSE, TRUE)))

plot_grid(plotlist = lapply(varpart.result, function(i) i[["plot"]]), nrow = length(tissues))

rm(form, run.varPart, varpart.result)
```

# Differential analysis (DA) of proteins within tissues

The table below shows the number of significant differentially expressed proteins between conditions within tissues. The values outside the parenethses are for a more relaxed Ï-value that corresponds to q-value of `r paste(significance.threshold.p.q.relaxed)` and the ones in the parentheses are for a more strict Ï-value that corresponds to q-value of `r paste(significance.threshold.p.q.strict)`. Ï-values are calculated as follows: -log~10~q*|logFC|. Here we use Ï-values as a measure of significance because they take into consideration both the FC and the q-value and because we have rather limited statistical power.

```{r within.tissues.differential.analysis, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
within.tissues.da.fun <- function(dia_, tissue_, coef_, correction_) {
  covars_ <- data.frame(id = tail(colnames(dia_), -1), stringsAsFactors = F) %>%
    left_join(select(covars, id, age, bmi, islets.purity, cit.min, pheno_3c), by = 'id') %>%
    filter(complete.cases(.))
  
  dia_ <- dia_ %>%
    column_to_rownames("AZpID") %>%
    select(all_of(covars_$id))
  
  if (tissue_ == "Pancreas") {
    design_ <- model.matrix(~ 0 + pheno_3c + bmi + age + islets.purity + cit.min, data = covars_)
    colnames(design_) <- c("ctrl", "pd", "t2d", "bmi", "age", "islets.purity", "cit.min")
  } else {
    design_ <- model.matrix(~ 0 + pheno_3c + bmi + age + cit.min, data = covars_)
    colnames(design_) <- c("ctrl", "pd", "t2d", "bmi", "age", "cit.min")
  }
  
  fit_ <- lmFit(dia_, design_)
  
  cm_ <- makeContrasts(
    t2dVSctrl = t2d - ctrl, # pair-wise t2d vs ctrl
    t2dVSpd = t2d - pd, # pair-wise t2d vs pd
    pdVSctrl = pd - ctrl, # pair-wise pd vs ctrl
    t2dVSctrl_pd = t2d - ((ctrl + pd) / 2), # t2d vs the average of ctrl and pd
    pd_t2dVSctrl = ((t2d + pd) / 2) - ctrl, # t2d vs the average of ctrl and pd
    levels = design_)
  
  fit2_ <- contrasts.fit(fit_, cm_)
  fit2_ <- eBayes(fit2_)
  
  tt <- topTable(fit = fit2_, coef = coef_, sort.by = "none", number = Inf, confint = T) %>%
    rownames_to_column("AZpID") %>%
    filter(complete.cases(.)) %>%
    mutate(Q.Value = qvalue::qvalue(p = P.Value)$qvalue,
           Pi.Q.Value = logFC * (-log10(Q.Value)),
           Pi.adjP.Value = logFC * (-log10(adj.P.Val))) %>%
    left_join(dia$index$index, by = 'AZpID') %>%
    mutate(sign = case_when(
               Pi.Q.Value >= significance.signs$pi[1] | Pi.Q.Value <= -significance.signs$pi[1] ~ significance.signs$sign[1],
               Pi.Q.Value >= significance.signs$pi[2] | Pi.Q.Value <= -significance.signs$pi[2] ~ significance.signs$sign[2],
               Pi.Q.Value >= significance.signs$pi[3] | Pi.Q.Value <= -significance.signs$pi[3] ~ significance.signs$sign[3],
               Pi.Q.Value >= significance.signs$pi[4] | Pi.Q.Value <= -significance.signs$pi[4] ~ significance.signs$sign[4],
               Pi.Q.Value >= significance.signs$pi[5] | Pi.Q.Value <= -significance.signs$pi[5] ~ significance.signs$sign[5],
               TRUE ~ significance.signs$sign[6])) %>%
    arrange(match(sign, significance.signs$sign), desc(abs(Pi.Q.Value))) %>%
    select(AZpID:Pi.adjP.Value, sign, everything())
  
  return(tt)
}

relaxed.significance <- significance.signs %>% filter(pi >= significance.threshold.pi) %>% pull(sign)
next.significance <- significance.signs %>% filter(pi > significance.threshold.pi) %>% pull(sign)

dia$results <- list()
dia$results$da <- list()
da.summary <- matrix(nrow = length(within.tissues.contrasts), ncol = length(tissues), dimnames = list(within.tissues.contrasts, tissues))
for (c in within.tissues.contrasts) {
  dia$results$da[[c]] <- lapply(tissues, function(t) within.tissues.da.fun(dia_ = dia$normalized.imputed[[t]], tissue_ = t, coef_ = c, correction_ = within.tissues.correction.method))
  names(dia$results$da[[c]]) <- tissues
  for (t in tissues) {
    tmp.res <- dia$results$da[[c]][[t]]
    sign.relaxed <- nrow(tmp.res %>% filter(sign %in% relaxed.significance))
    sign.next <- nrow(tmp.res %>% filter(sign %in% next.significance))
    da.summary[c, t] <- paste(sign.relaxed, " (", sign.next, ")", sep = "")
  }
}

rm(t, c, tmp.res, sign.relaxed, sign.next, relaxed.significance, next.significance, tt.sign.fc, within.tissues.da.fun, fix.uniprot.ids)

# write DA results in xlsx files
for (i in names(dia$results$da)) {
  comp = dia$results$da[[i]]
  if (file.exists(paste(output.directory, "within_tissues_results_da_", i, ".xlsx", sep = ""))) {
    file.remove(paste(output.directory, "within_tissues_results_da_", i, ".xlsx", sep = ""))
  }
  lapply(seq_along(comp), function(j) write.to.excel(df_ = comp[[j]], file_ = paste(output.directory, "within_tissues_results_da_", i, ".xlsx", sep = ""),
                                                     sheet_ = names(comp)[j], append_ = ifelse(j == tissues[1], FALSE, TRUE)))
}
rm(comp, i)

da.summary <- as.data.frame(da.summary) %>%
  rownames_to_column("diff") %>%
  left_join(data.frame(diff = within.tissues.contrasts, my.diff = names(within.tissues.contrasts), stringsAsFactors = F), by = 'diff') %>%
  select(diff, my.diff, everything()) %>%
  select(-diff) %>%
  rename("Contrasts" = "my.diff")
if (file.exists(paste(output.directory, "within_tissues_da_results_summary.xlsx", sep = ""))) {
  file.remove(paste(output.directory, "within_tissues_da_results_summary.xlsx", sep = ""))
}
write.to.excel(df_ = da.summary, file_ = paste(output.directory, "within_tissues_da_results_summary.xlsx", sep = ""), sheet_ = "Sheet1", append_ = F)
kable(da.summary, align = 'r') %>% 
  kable_styling(full_width = F, bootstrap_options = "striped") %>%
  column_spec(1, bold = T, border_right = T)
rm(da.summary)
```

# Gene-set enrichment analysis (GSEA)

## GSEA for gene ontology (GO) terms in tissue-specific proteins

The plot below shows the top-5 most enriched GO terms for proteins identified exclusively in one single tissue.

```{r tissue.specific.gsea, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=17, fig.height=12}
# hypergeometric test
hg.test <- function(num.in.term, num.of.set, num.in.term.from.universe, num.of.universe, which.tail) {
  if (which.tail == "UPPER") {
    num.in.term <- num.in.term - 1
    lower.tail <- FALSE
  } else if (which.tail == "LOWER") {
    lower.tail <- TRUE
  }
  
  num.not.in.go.term.from.universe <- num.of.universe - num.in.term.from.universe
  phyper(num.in.term,
         num.in.term.from.universe,
         num.not.in.go.term.from.universe,
         num.of.set, lower.tail = lower.tail)
}

hg.gsea <- function(tissue_, terms.file_) {
  terms.set <- fgsea::gmtPathways(terms.file_)
  
  num.in.term.from.universe <- sapply(terms.set, function(i) sum(dia$index$AZpID$Gene %in% i))
  
  the.set <- dia$index$AZpID %>%
    mutate(cnt = select(., starts_with("AZpID.")) %>% is.na(.) %>% `!` %>% rowSums(., na.rm = T)) %>%
    select(AZpID, Gene, ends_with(tissue_), cnt) %>%
    rename("in.tissue" = paste("AZpID", tissue_, sep = ".")) %>%
    filter(cnt == 1 & !is.na(in.tissue)) %>%
    pull(Gene)
  
  num.in.term <- sapply(terms.set, function(i) sum(the.set %in% i))
  proteins.in.term <- sapply(lapply(terms.set, function(i) the.set[the.set %in% i]), paste, collapse = ",")
  
  num.in.term <- num.in.term[num.in.term.from.universe > 0]
  proteins.in.term <- proteins.in.term[num.in.term.from.universe > 0]
  num.in.term.from.universe <- num.in.term.from.universe[num.in.term.from.universe > 0]
  
  p.values <- sapply(seq_along(num.in.term.from.universe), function(i) hg.test(num.in.term = num.in.term[i], num.of.set = length(the.set), num.in.term.from.universe = num.in.term.from.universe[i], num.of.universe = nrow(dia$index$AZpID), which.tail = "UPPER"))
  names(p.values) <- names(num.in.term)
  
  q.values <- qvalue::qvalue(p = p.values)$qvalue
  
  return(data.frame(tissue = tissue_,
                    p.value = p.values,
                    q.value = q.values,
                    in.set = num.in.term,
                    protein.id = proteins.in.term, stringsAsFactors = F) %>%
           rownames_to_column("pathway"))
}

dia$results$tissue.specific.gsea <- bind_rows(lapply(tissues, hg.gsea, terms.file_ = GOall.file)) %>%
  group_by(pathway) %>%
  filter(any(q.value <= significance.threshold.p.q.strict)) %>%
  mutate(log.p.adj = -log10(q.value)) %>%
  mutate(term.name = unlist(lapply(strsplit(pathway, "|", fixed = T), head, 1)),
         term.id = unlist(lapply(strsplit(pathway, "|", fixed = T), tail, 1)),
         term.db = unlist(lapply(lapply(strsplit(pathway, "|", fixed = T), head, 2), tail, 1)),
         term.name = ifelse(term.db %in% biological.terms.files$combo, term.name, paste(term.name, substr(term.db, 1, 1))),
         term.db = ifelse(term.db %in% biological.terms.files$combo, term.db, "Pathway")) %>%
  arrange(term.db, tissue, desc(log.p.adj)) %>%
  left_join(as.data.frame(sapply(fgsea::gmtPathways(GOall.file), length)) %>% magrittr::set_colnames(c("set")) %>% rownames_to_column("name"), by = c("pathway" = "name"))

dia$results$tissue.specific.gsea %>%
  mutate(log.p.adj = replace(log.p.adj, log.p.adj > 10, 10)) %>%
  group_by(term.db, tissue) %>%
  arrange(desc(log.p.adj)) %>%
  slice(1:5) %>%
  ggplot(aes(x = tissue, y = term.name, fill = log.p.adj)) +
  geom_tile() +
  scale_fill_gradient2(name = expression('-log'[10]*'q')) + 
  scale_x_discrete(expand = c(0, 0), drop = TRUE) + 
  facet_wrap(~ term.db, ncol = 4) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12),
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.border = element_rect(color = "grey80", fill = NA, size = .8))

rm(hg.gsea)
```

## GSEA for gene ontology (GO) terms in tissue-shared proteins

The plot below shows the top-5 most enriched GO terms for proteins shared among 2 or more tissues.

```{r tissue.shared.gsea, echo=FALSE, message=FALSE, warning=FALSE, results='hide', fig.show='asis', fig.keep='last', fig.width=19, fig.height=25}
hg.gsea <- function(the.set_, tissue_, terms.file_) {
  terms.set <- fgsea::gmtPathways(terms.file_)
  
  num.in.term.from.universe <- sapply(terms.set, function(i) sum(dia$index$AZpID$Gene %in% i))
  
  num.in.term <- sapply(terms.set, function(i) sum(the.set_ %in% i))
  proteins.in.term <- sapply(lapply(terms.set, function(i) the.set_[the.set_ %in% i]), paste, collapse = ",")
  
  num.in.term <- num.in.term[num.in.term.from.universe > 0]
  proteins.in.term <- proteins.in.term[num.in.term.from.universe > 0]
  num.in.term.from.universe <- num.in.term.from.universe[num.in.term.from.universe > 0]
  
  p.values <- sapply(seq_along(num.in.term.from.universe), function(i) hg.test(num.in.term = num.in.term[i],
                                                                               num.of.set = length(the.set_),
                                                                               num.in.term.from.universe = num.in.term.from.universe[i],
                                                                               num.of.universe = nrow(dia$index$AZpID),
                                                                               which.tail = "UPPER"))
  names(p.values) <- names(num.in.term)
  
  q.values <- qvalue::qvalue(p = p.values)$qvalue
  
  return(data.frame(tissue = tissue_,
                    p.value = p.values,
                    q.value = q.values,
                    in.set = num.in.term,
                    protein.id = proteins.in.term, stringsAsFactors = F) %>%
           rownames_to_column("pathway"))
}

pa.set <- dia$index$AZpID %>%
  mutate_at(.vars = vars(starts_with("AZpID.")), .funs = list(~unlist(lapply(strsplit(x = ., split = ".", fixed = T), head, 1)))) %>%
  mutate(cnt = select(., starts_with("AZpID.")) %>% is.na(.) %>% `!` %>% rowSums(x = ., na.rm = T)) %>%
  unite("AZpID.name", starts_with("AZpID."), sep = "-", remove = T, na.rm = T) %>%
  filter(cnt > 1) %>%
  arrange(cnt, AZpID.name) %>%
  select(Gene, AZpID.name)

dia$results$tissue.shared.gsea <- bind_rows(lapply(unique(pa.set$AZpID.name), function(i) hg.gsea(the.set_ = pa.set %>% filter(AZpID.name == i) %>% pull(Gene), tissue_ = i, terms.file_ = GOall.file))) %>%
  group_by(pathway) %>%
  filter(any(q.value <= significance.threshold.p.q.strict)) %>%
  mutate(log.p.adj = -log10(q.value)) %>%
  mutate(term.name = unlist(lapply(strsplit(pathway, "|", fixed = T), head, 1)),
         term.id = unlist(lapply(strsplit(pathway, "|", fixed = T), tail, 1)),
         term.db = unlist(lapply(lapply(strsplit(pathway, "|", fixed = T), head, 2), tail, 1)),
         term.name = ifelse(term.db %in% biological.terms.files$combo, term.name, paste(term.name, substr(term.db, 1, 1))),
         term.db = ifelse(term.db %in% biological.terms.files$combo, term.db, "Pathway")) %>%
  arrange(term.db, tissue, desc(log.p.adj)) %>%
  left_join(as.data.frame(sapply(fgsea::gmtPathways(GOall.file), length)) %>% magrittr::set_colnames(c("set")) %>% rownames_to_column("name"), by = c("pathway" = "name"))

dia$results$tissue.shared.gsea %>%
  mutate(log.p.adj = replace(log.p.adj, log.p.adj > 10, 10)) %>%
  group_by(term.db, tissue) %>%
  arrange(desc(log.p.adj)) %>%
  slice(1:5) %>%
  ggplot(aes(x = tissue, y = term.name, fill = log.p.adj)) +
  geom_tile() +
  scale_fill_gradient2(name = expression('-log'[10]*'q')) + 
  scale_x_discrete(expand = c(0, 0), drop = TRUE) + 
  facet_wrap(~ term.db, ncol = 4) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12),
        axis.title = element_blank(),
        axis.ticks = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.border = element_rect(color = "grey80", fill = NA, size = .8))

rm(hg.gsea, hg.test, pa.set)
```

## GSEA for gene ontology (GO) terms based on DA

The table below summarizes number of enriched GO terms and pathways for each tissue. We have used GO term for biological processes (bp), molecular function (mf) and cellular components (cc), and pathways from KEGG and Reactome. Each row stands for a different pair-wise comparison of phenotypes. Significance in this cases is measured with q-values. The values outside the parentheses stand for q-values less than `r paste(significance.threshold.p.q.strict)` and the ones within the parentheses for q-values less than `r significance.threshold.p.q.relaxed / 10`.

```{r within.tissue.enrichment.analysis, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
run.gsea <- function(limma.tt_, terms.file_, significance.threshold_, min.num.of.proteins_, max.num.of.proteins_, num.of.permut_, mode_, snow.workers_) {
  library(dplyr)
  library(tibble)
  
  if (!(mode_ %in% c("fgsea", "camera", "fgsea+camera"))) {
    stop("Error: the mode in the function run.gsea is incorrect!")
  }
  
  # list.of.proteins_ <- setNames(object = limma.tt_$logFC, nm = limma.tt_$Gene)
  list.of.proteins_ <- setNames(object = limma.tt_$Pi.Q.Value, nm = limma.tt_$Gene)
  
  # take care of the input file
  if (any(duplicated(names(list.of.proteins_)))) {
    list.of.proteins_ <- list.of.proteins_[!duplicated(names(list.of.proteins_))]
  }
  
  if (!all(order(list.of.proteins_, decreasing = TRUE) == 1:length(list.of.proteins_))) {
    list.of.proteins_ <- sort(list.of.proteins_, decreasing = TRUE)
  }
  
  # prepare the set of terms to analyze
  terms.set <- fgsea::gmtPathways(terms.file_)
  
  # run and filter fgsea
  if (stringr::str_detect(string = mode_, pattern = "fgsea")) {
    fgsea.result <- fgsea::fgsea(pathways = terms.set, stats = list.of.proteins_, minSize = min.num.of.proteins_, maxSize = max.num.of.proteins_, nperm = num.of.permut_, BPPARAM = BiocParallel::SnowParam(workers = snow.workers_, type = "SOCK")) %>%
      as.data.frame() %>%
      mutate(leadingEdge = sapply(leadingEdge, paste, collapse = ",")) %>%
      filter(complete.cases(.)) %>%
      mutate(padj = qvalue::qvalue(pval)$qvalue)
  } else {
    fgsea.result <- fgsea::fgsea(pathways = terms.set, stats = list.of.proteins_, minSize = min.num.of.proteins_, maxSize = max.num.of.proteins_, nperm = 3) %>%
      mutate(leadingEdge = sapply(leadingEdge, paste, collapse = ","))
  }
  
  # run camera
  if (stringr::str_detect(string = mode_, pattern = "camera")) {
    camera.result <- limma::cameraPR(statistic = list.of.proteins_, index = terms.set, inter.gene.cor = 0.01, sort = T) %>%
      rownames_to_column("pathway")
  }
  
  if (mode_ == "fgsea") {
    result.merge <- fgsea.result %>%
      mutate(significant = ifelse(padj <= significance.threshold_, TRUE, FALSE),
             allEdges = unlist(lapply(pathway, function(i) paste(terms.set[[i]][terms.set[[i]] %in% names(list.of.proteins_)], collapse = ","))))
  } else if (mode_ == "camera") {
    result.merge <- camera.result %>%
      mutate(pval = PValue,
             padj = FDR,
             ES = ifelse(Direction == "Up", 1, -1),
             NES = stat.mean,
             nMoreExtreme = 0,
             size = NGenes) %>%
      left_join(select(fgsea.result, pathway, leadingEdge), by = 'pathway') %>%
      select(pathway, pval, padj, ES, NES, nMoreExtreme, size, leadingEdge) %>%
      mutate(significant = ifelse(padj <= significance.threshold_, TRUE, FALSE),
             allEdges = unlist(lapply(pathway, function(i) paste(terms.set[[i]][terms.set[[i]] %in% names(list.of.proteins_)], collapse = ","))))
  } else if (mode_ == "fgsea+camera") {
    result.merge <- left_join(fgsea.result, camera.result, by = 'pathway') %>%
      filter((NES < 0 & Direction == "Down") | (NES > 0 & Direction == "Up")) %>%
      mutate(significant = case_when(
        FDR <= significance.threshold_ & padj <= significance.threshold_ ~ TRUE,
        TRUE ~ FALSE),
        pval = PValue,
        padj = FDR) %>%
      select(all_of(colnames(fgsea.result)), significant) %>%
      mutate(allEdges = unlist(lapply(pathway, function(i) paste(terms.set[[i]][terms.set[[i]] %in% names(list.of.proteins_)], collapse = ","))))
  }
  
  if (nrow(result.merge) == 0) {
    result.merge.fin <- result.merge
  } else {
    result.merge.fin <- result.merge %>%
      mutate(Enrichment = ifelse(NES > 0, "UP", "DOWN"),
             term.name = unlist(lapply(strsplit(x = pathway, split = "|", fixed = T), head, 1)),
             term.db = unlist(lapply(lapply(strsplit(x = pathway, split = "|", fixed = T), head, 2), tail, 1)),
             term.id = unlist(lapply(strsplit(x = pathway, split = "|", fixed = T), tail, 1)),
             absNES = abs(NES),
             logpadj = -log10(padj)) %>%
      group_by(Enrichment) %>%
      arrange(desc(logpadj)) %>%
      ungroup() %>%
      group_by(term.name) %>%
      arrange(desc(logpadj)) %>%
      slice(1) %>%
      ungroup() %>%
      arrange(Enrichment, logpadj) %>%
      mutate(term.name = factor(x = term.name, levels = term.name),
             Enrichment = factor(x = Enrichment, levels = sort(unique(Enrichment))))
  }
  
  return(result.merge.fin)
}

gsea.summary <- lapply(biological.terms.files$combo, function(i) matrix(nrow = length(within.tissues.contrasts), ncol = length(tissues), dimnames = list(within.tissues.contrasts, tissues)))
names(gsea.summary) <- biological.terms.files$combo

for (terms.combo in biological.terms.files$combo) {
  res.name <- paste("within.tissue.gsea", terms.combo, sep = ".")
  dia$results[[res.name]] <- list()
  for(diff.analysis in names(dia$results$da)) {
    dia$results[[res.name]][[diff.analysis]] <- lapply(dia$results$da[[diff.analysis]], run.gsea,
                                                       terms.file_ = filter(biological.terms.files, combo == terms.combo) %>% slice(1) %>% pull(file),
                                                       significance.threshold_ = significance.threshold.p.q.strict,
                                                       min.num.of.proteins_ = min.num.of.proteins.in.term,
                                                       max.num.of.proteins_ = max.num.of.proteins.term,
                                                       num.of.permut_ = num.of.permutations,
                                                       mode_ = "fgsea+camera",
                                                       snow.workers_ = num.of.cores.total)
    gsea.summary[[terms.combo]][diff.analysis,] <- unname(unlist(lapply(dia$results[[res.name]][[diff.analysis]], function(i) paste(i %>% filter(logpadj >= -log10(significance.threshold.p.q.strict) & significant == TRUE) %>% nrow(), " (", i %>% filter(logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE) %>% nrow(), ")", sep = ""))))
  }
  s <- lapply(dia$results[[res.name]], function(i) i[lapply(i, nrow) > 0])
  res.name <- paste("within.tissue.gsea", terms.combo, "df", sep = ".")
  dia$results[[res.name]] <- bind_rows(lapply(seq_along(s), function(j) bind_rows(lapply(seq_along(s[[j]]), function(i) mutate(s[[j]][[i]], tissue = names(s[[j]])[i]) %>% mutate_all(as.character) %>% map_df(~parse_guess(.)))) %>% mutate(diff = names(s)[j])))
}

rm(terms.combo, diff.analysis, run.gsea, s, res.name)

gsea.summary <- plyr::join_all(dfs = lapply(seq_along(gsea.summary), function(i) gsea.summary[[i]] %>% as.data.frame() %>% rownames_to_column("diff") %>% rename_at(.vars = vars(-diff), .funs = list(~ paste(., names(gsea.summary)[i], sep = ".")))), by = 'diff', type = 'full') %>%
  left_join(data.frame(diff = within.tissues.contrasts, my.diff = names(within.tissues.contrasts), stringsAsFactors = F), by = 'diff') %>%
  select(diff, my.diff, everything()) %>%
  select(-diff) %>%
  rename("Contrasts" = "my.diff")
if (file.exists(paste(output.directory, "within_tissues_gsea_results_summary.xlsx", sep = ""))) {
  file.remove(paste(output.directory, "within_tissues_gsea_results_summary.xlsx", sep = ""))
}
write.to.excel(df_ = gsea.summary, file_ = paste(output.directory, "within_tissues_gsea_results_summary.xlsx", sep = ""), sheet_ = "Sheet1", append_ = F)
kable(gsea.summary, align = 'r', col.names = c(head(colnames(gsea.summary), 1), unlist(lapply(strsplit(x = tail(colnames(gsea.summary), -1), split = ".", fixed = T), head, 1)))) %>% 
  kable_styling(full_width = F, bootstrap_options = "striped", font_size = 6.5) %>%
  column_spec(1, bold = T, border_right = T) %>%
  add_header_above(setNames(object = c(1, rep(length(tissues), nrow(biological.terms.files))), nm = c(" ", biological.terms.files$combo)))
rm(gsea.summary)
```

## Visualization of GSEA for GO terms

### GSEA for GO biological processes

The plots below show enriched GO terms for biological processes for each tissue (q<=`r paste(significance.threshold.p.q.strictest)`). The 3 first panels show pair-wise enrichments for comparisons of the 3 phenotypic classes _CTRL_, _PD_ and _T2D_, while the 2 last ones show enrichemns for comparisons between _CTRL+PD_ versus _T2D_, and _PD+T2D_ versus _CTRL_, essentially exploring how much the landscape of GO-enriched terms changes due to shifting PD.

```{r within.tissue.enrichment.analysis.plotGObp, echo=FALSE, message=FALSE, warning=FALSE, results='asis', fig.show='asis', fig.keep='last', fig.width=17, fig.height=12}
my.combo <- "GObp"

goslim <- read.delim(file = GOslim.file, header = T, sep = "\t", stringsAsFactors = F)
goslim.cols <- setNames(object = unique(filter(goslim, namespace_custom == my.combo) %>% arrange(name) %>% pull(color)),
                        nm = unique(filter(goslim, namespace_custom == my.combo) %>% arrange(name) %>% pull(name)))

if (nrow(filter(dia$results$within.tissue.gsea.GObp.df, logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE)) > 0) {
  if(file.exists(paste(output.directory, "within_tissues_gsea_results.xlsx", sep = ""))) {
    file.remove(paste(output.directory, "within_tissues_gsea_results.xlsx", sep = ""))
  }
  dia$results$within.tissue.gsea.GObp.df %>%
    left_join(filter(goslim, namespace_custom == my.combo), by = c("term.id" = "slimsFromId")) %>%
    filter(logpadj >= -log10(significance.threshold.p.q.strict) & significant == TRUE) %>%
    select(pathway, term.name:term.id, slimsToIds:name, pval, padj, logpadj, ES, NES, Enrichment, tissue, diff, size, leadingEdge, significant) %>%
    rename("pathway full name" = "pathway", "pathway name" = "term.name", "pathway database" = "term.db", "pathway id" = "term.id", "GOslim pathway id" = "slimsToIds", "GOslim pathway name" = "name", "p value" = "pval", "q value" = "padj", "log q value" = "logpadj", "enrichment score" = "ES", "normalized enrichment score" = "NES", "direction" = "Enrichment", "comparison"= "diff", "overlap" = "size", "protein ids" = "leadingEdge", "Significant in all" = "significant") %>%
    left_join(data.frame(comparison = within.tissues.contrasts, c.names = names(within.tissues.contrasts), stringsAsFactors = F), by = 'comparison') %>%
    select(-comparison) %>%
    rename("comparison"= "c.names") %>%
    select("pathway full name":"tissue", "comparison", everything()) %>%
    write.to.excel(df_ = ., file_ = paste(output.directory, "within_tissues_gsea_results.xlsx", sep = ""), sheet_ = "GObp", append_ = F)
  
  dia$results$within.tissue.gsea.GObp.df %>%
    left_join(filter(goslim, namespace_custom == my.combo), by = c("term.id" = "slimsFromId")) %>%
    filter(logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE) %>%
    arrange(name) %>%
    mutate(Enrichment = factor(Enrichment, levels = c("DOWN", "UP")),
           name = factor(x = name, levels = sort(unique(name))),
           term.name = factor(x = term.name, levels = unique(term.name)),
           diff = factor(x = diff, levels = within.tissues.contrasts)) %>%
    ggplot(aes(x = tissue, y = term.name, shape = Enrichment, color = name, fill = name)) +
    scale_color_manual(name = "GOslim", values = goslim.cols, na.value = 'grey50') +
    scale_shape_manual(name = "Direction", values = c("DOWN" = 25, "UP" = 24)) +
    scale_fill_manual(name = "GOslim", values = goslim.cols, na.value = 'grey50') + 
    geom_point(aes(size = logpadj)) +
    facet_wrap(~ diff, labeller = labeller(diff = setNames(object = names(within.tissues.contrasts), nm = within.tissues.contrasts)), ncol = length(within.tissues.contrasts)) +
    guides(size = guide_legend(title = expression('-log'[10]*'q'))) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12),
          axis.title = element_blank(),
          axis.ticks = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.border = element_rect(color = "grey80", fill = NA, size = .8))
} else {
  print("No significant GObp terms!")
}

rm(goslim, goslim.cols, my.combo)
```

### GSEA for GO cellular components

The plots below show enriched GO terms for cellular components for each tissue (q<=`r paste(significance.threshold.p.q.strictest)`). The 3 first panels show pair-wise enrichments for comparisons of the 3 phenotypic classes _CTRL_, _PD_ and _T2D_, while the 2 last ones show enrichemns for comparisons between _CTRL+PD_ versus _T2D_, and _PD+T2D_ versus _CTRL_, essentially exploring how much the landscape of GO-enriched terms changes due to shifting PD.

```{r within.tissue.enrichment.analysis.plotGOcc, echo=FALSE, message=FALSE, warning=FALSE, results='asis', fig.show='asis', fig.keep='last', fig.width=17, fig.height=12}
my.combo <- "GOcc"

goslim <- read.delim(file = GOslim.file, header = T, sep = "\t", stringsAsFactors = F)
goslim.cols <- setNames(object = unique(filter(goslim, namespace_custom == my.combo) %>% arrange(name) %>% pull(color)), nm = unique(filter(goslim, namespace_custom == my.combo) %>% arrange(name) %>% pull(name)))

if (nrow(filter(dia$results$within.tissue.gsea.GOcc.df, logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE)) > 0) {
  dia$results$within.tissue.gsea.GOcc.df %>%
    left_join(filter(goslim, namespace_custom == my.combo), by = c("term.id" = "slimsFromId")) %>%
    filter(logpadj >= -log10(significance.threshold.p.q.strict) & significant == TRUE) %>%
    select(pathway, term.name:term.id, slimsToIds:name, pval, padj, logpadj, ES, NES, Enrichment, tissue, diff, size, leadingEdge, significant) %>%
    rename("pathway full name" = "pathway", "pathway name" = "term.name", "pathway database" = "term.db", "pathway id" = "term.id", "GOslim pathway id" = "slimsToIds", "GOslim pathway name" = "name", "p value" = "pval", "q value" = "padj", "log q value" = "logpadj", "enrichment score" = "ES", "normalized enrichment score" = "NES", "direction" = "Enrichment", "comparison"= "diff", "overlap" = "size", "protein ids" = "leadingEdge", "Significant in all" = "significant") %>%
    left_join(data.frame(comparison = within.tissues.contrasts, c.names = names(within.tissues.contrasts), stringsAsFactors = F), by = 'comparison') %>%
    select(-comparison) %>%
    rename("comparison"= "c.names") %>%
    select("pathway full name":"tissue", "comparison", everything()) %>%
    write.to.excel(df_ = ., file_ = paste(output.directory, "within_tissues_gsea_results.xlsx", sep = ""), sheet_ = "GOcc", append_ = T)
  
  dia$results$within.tissue.gsea.GOcc.df %>%
    left_join(filter(goslim, namespace_custom == my.combo), by = c("term.id" = "slimsFromId")) %>%
    filter(logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE) %>%
    arrange(name) %>%
    mutate(Enrichment = factor(Enrichment, levels = c("DOWN", "UP")),
           name = factor(x = name, levels = sort(unique(name))),
           term.name = factor(x = term.name, levels = unique(term.name)),
           diff = factor(x = diff, levels = within.tissues.contrasts)) %>%
    ggplot(aes(x = tissue, y = term.name, shape = Enrichment, color = name, fill = name)) +
    scale_color_manual(name = "GOslim", values = goslim.cols, na.value = 'grey50') +
    scale_shape_manual(name = "Direction", values = c("DOWN" = 25, "UP" = 24)) +
    scale_fill_manual(name = "GOslim", values = goslim.cols, na.value = 'grey50') + 
    geom_point(aes(size = logpadj)) +
    facet_wrap(~ diff, labeller = labeller(diff = setNames(object = names(within.tissues.contrasts), nm = within.tissues.contrasts)), ncol = length(within.tissues.contrasts)) +
    guides(size = guide_legend(title = expression('-log'[10]*'q'))) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12),
          axis.title = element_blank(),
          axis.ticks = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.border = element_rect(color = "grey80", fill = NA, size = .8))
} else {
  print("No significant GOcc terms!")
}

rm(goslim, goslim.cols, my.combo)
```

### GSEA for GO molecular functions

The plots below show enriched GO terms for molecular functions for each tissue (q<=`r paste(significance.threshold.p.q.strictest)`). The 3 first panels show pair-wise enrichments for comparisons of the 3 phenotypic classes _CTRL_, _PD_ and _T2D_, while the 2 last ones show enrichemns for comparisons between _CTRL+PD_ versus _T2D_, and _PD+T2D_ versus _CTRL_, essentially exploring how much the landscape of GO-enriched terms changes due to shifting PD.

```{r within.tissue.enrichment.analysis.plotGOmf, echo=FALSE, message=FALSE, warning=FALSE, results='asis', fig.show='asis', fig.keep='last', fig.width=17, fig.height=12}
my.combo <- "GOmf"

goslim <- read.delim(file = GOslim.file, header = T, sep = "\t", stringsAsFactors = F)
goslim.cols <- setNames(object = unique(filter(goslim, namespace_custom == my.combo) %>% arrange(name) %>% pull(color)),
                        nm = unique(filter(goslim, namespace_custom == my.combo) %>% arrange(name) %>% pull(name)))

if (nrow(filter(dia$results$within.tissue.gsea.GOmf.df, logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE)) > 0) {
  dia$results$within.tissue.gsea.GOmf.df %>%
    left_join(filter(goslim, namespace_custom == my.combo), by = c("term.id" = "slimsFromId")) %>%
    filter(logpadj >= -log10(significance.threshold.p.q.strict) & significant == TRUE) %>%
    select(pathway, term.name:term.id, slimsToIds:name, pval, padj, logpadj, ES, NES, Enrichment, tissue, diff, size, leadingEdge, significant) %>%
    rename("pathway full name" = "pathway", "pathway name" = "term.name", "pathway database" = "term.db", "pathway id" = "term.id", "GOslim pathway id" = "slimsToIds", "GOslim pathway name" = "name", "p value" = "pval", "q value" = "padj", "log q value" = "logpadj", "enrichment score" = "ES", "normalized enrichment score" = "NES", "direction" = "Enrichment", "comparison"= "diff", "overlap" = "size", "protein ids" = "leadingEdge", "Significant in all" = "significant") %>%
    left_join(data.frame(comparison = within.tissues.contrasts, c.names = names(within.tissues.contrasts), stringsAsFactors = F), by = 'comparison') %>%
    select(-comparison) %>%
    rename("comparison"= "c.names") %>%
    select("pathway full name":"tissue", "comparison", everything()) %>%
    write.to.excel(df_ = ., file_ = paste(output.directory, "within_tissues_gsea_results.xlsx", sep = ""), sheet_ = "GOmf", append_ = T)
  
  dia$results$within.tissue.gsea.GOmf.df %>%
    left_join(filter(goslim, namespace_custom == my.combo), by = c("term.id" = "slimsFromId")) %>%
    filter(logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE) %>%
    arrange(name) %>%
    mutate(Enrichment = factor(Enrichment, levels = c("DOWN", "UP")),
           name = factor(x = name, levels = sort(unique(name))),
           term.name = factor(x = term.name, levels = unique(term.name)),
           diff = factor(x = diff, levels = within.tissues.contrasts)) %>%
    ggplot(aes(x = tissue, y = term.name, shape = Enrichment, color = name, fill = name)) +
    scale_color_manual(name = "GOslim", values = goslim.cols, na.value = 'grey50') +
    scale_shape_manual(name = "Direction", values = c("DOWN" = 25, "UP" = 24)) +
    scale_fill_manual(name = "GOslim", values = goslim.cols, na.value = 'grey50') + 
    geom_point(aes(size = logpadj)) +
    facet_wrap(~ diff, labeller = labeller(diff = setNames(object = names(within.tissues.contrasts), nm = within.tissues.contrasts)), ncol = length(within.tissues.contrasts)) +
    guides(size = guide_legend(title = expression('-log'[10]*'q'))) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12),
          axis.title = element_blank(),
          axis.ticks = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.border = element_rect(color = "grey80", fill = NA, size = .8))
} else {
  print("No significant GObp terms!")
}

rm(goslim, goslim.cols, my.combo)
```

## Visualization of GSEA for pathways based on DA

The plots below show enriched pathways for each tissue (q<=`r paste(significance.threshold.p.q.strictest)`). The plot on the left hand-side shows enrichemnt for pair-wise comparisons of _CTRL+PD_ versus _T2D_, and _PD+T2D_ versus _CTRL_, essentially exploring how much the landscape of enriched pathways changes due to shifting PD. The right-hand side plots show pair-wise comparisons of the 3 phenotypic classes _CTRL_, _PD_ and _T2D_.

```{r within.tissue.enrichment.analysis.plotPathway, echo=FALSE, message=FALSE, warning=FALSE, results='asis', fig.show='asis', fig.keep='last', fig.width=22, fig.height=9}
if (nrow(filter(dia$results$within.tissue.gsea.Pathway.df, logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE)) > 0) {
  dia$results$within.tissue.gsea.Pathway.df %>%
    filter(logpadj >= -log10(significance.threshold.p.q.strict) & significant == TRUE) %>%
    select(pathway, term.name:term.id, pval, padj, logpadj, ES, NES, Enrichment, tissue, diff, size, leadingEdge, significant) %>%
    rename("pathway full name" = "pathway", "pathway name" = "term.name", "pathway database" = "term.db", "pathway id" = "term.id", "p value" = "pval", "q value" = "padj", "log q value" = "logpadj", "enrichment score" = "ES", "normalized enrichment score" = "NES", "direction" = "Enrichment", "comparison"= "diff", "overlap" = "size", "protein ids" = "leadingEdge", "Significant in all" = "significant") %>%
    left_join(data.frame(comparison = within.tissues.contrasts, c.names = names(within.tissues.contrasts), stringsAsFactors = F), by = 'comparison') %>%
    select(-comparison) %>%
    rename("comparison"= "c.names") %>%
    select("pathway full name":"tissue", "comparison", everything()) %>%
    write.to.excel(df_ = ., file_ = paste(output.directory, "within_tissues_gsea_results.xlsx", sep = ""), sheet_ = "Pathway", append_ = T)
  
  dia$results$within.tissue.gsea.Pathway.df %>%
    filter(logpadj >= -log10(significance.threshold.p.q.strictest) & significant == TRUE) %>%
    arrange(tissue, diff, term.db) %>%
    mutate(term.name2 = paste(term.name, substr(x = term.db, start = 1, stop = 1))) %>%
    mutate(Enrichment = factor(Enrichment, levels = c("DOWN", "UP")),
           term.name2 = factor(x = term.name2, levels = unique(term.name2)),
           diff = factor(x = diff, levels = within.tissues.contrasts)) %>%
    ggplot(aes(x = tissue, y = term.name2, color = Enrichment)) +
    scale_color_manual(name = "Direction", values = c("DOWN" = '#426E86FF', "UP" = '#F9BA32FF')) +
    geom_point(aes(size = logpadj)) +
    facet_wrap(~ diff, labeller = labeller(diff = setNames(object = names(within.tissues.contrasts), nm = within.tissues.contrasts)), ncol = length(within.tissues.contrasts)) +
    guides(size = guide_legend(title = expression('-log'[10]*'q'))) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size = 12),
          axis.title = element_blank(),
          axis.ticks = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.border = element_rect(color = "grey80", fill = NA, size = .8))
} else {
  print("No significant GObp terms!")
}
```

# Saving the relevant variables and session info

```{r cleaning.up, echo=FALSE, warning=FALSE, message=FALSE, results='asis'}
rm(write.to.excel)
paste("Ended at:", format(Sys.time(), '%A, %d %B %Y, %H:%M'))
```

```{r saving.the.session, echo=TRUE, warning=FALSE, message=FALSE, results='hide'}
save.image(file = paste(output.directory, "proteomics_differential_analysis.RData", sep = ""), compress = "gzip", safe = TRUE)
```

```{r session.info, echo=FALSE, warning=FALSE, message=FALSE, results='asis'}
sessionInfo()
```
